---
title: "<Project> RNAseq"
author: "Miles Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 1500px !important;
      width: 1500px !important;
    }
    body {
      max-width: 1500px !important;
    }
```

```{r setup, include=FALSE}
options(width = 1200)
knitr::opts_chunk$set(echo = FALSE)
loadd(annotation_info,
  annotated_module_scores,
  annotated_modules,
  sle_module_scores,
  ldg_module_scores,
  dge_qc,
  fit,
  qlf,
  v,
  group_pal,
  ifn_scores,
  inflammation_scores,
  ifn_scores_with_viral,
  inflammation_scores_with_viral,
  ISGs,
  ldg_scores,
  md,
  md_cat_data,
  md_num_data,
  module_ISGs,
  module_scores,
  module_tbl,
  pca_qc,
  pca_plot,
  pca_plot2,
  pca_res,
  removed_outliers,
  res,
  sample_dendrogram,
  sample_dists,
  sampleDistMatrix,
  up_table,
  down_table,
  top_up,
  top_down,
  tx_files,
  umap_plot2,
  umap_plot,
  umap_results,
  viral_transcripts,
  v)
```

# Import of all SLE and control RNAseq data

Overview of sample characteristics:

```{r fig.height=6, fig.width=6}
show_plot(md_cat_data, high_cardinality=1, col_palette = 3)
```

```{r}
show_plot(md_num_data, col_palette = 3)
```

Breakdown of the source of samples:

```{r}
md %>%
  tabyl(project) %>%
  adorn_pct_formatting(digits = 2) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, bold = TRUE, border_right = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = FALSE, width = "10em")

md %>%
  tabyl(study_group) %>%
  adorn_pct_formatting(digits = 2) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, bold = TRUE, border_right = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = FALSE, width = "10em")

md %>%
  tabyl(disease_class) %>%
  adorn_pct_formatting(digits = 2) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, bold = TRUE, border_right = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = FALSE, width = "10em")

```

```{r}
md %>% 
  tabyl(run_id, disease_class) %>%
  adorn_totals("col") %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, border_right = TRUE, bold = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = TRUE, width = "10em") %>%
  column_spec(4, border_right = FALSE, width = "10em")
```

```{r echo=FALSE}
message(str_glue("{length(md$sample_name)} samples selected for analysis"))
message(str_glue("{length(names(tx_files))} files found"))
if (!all(names(tx_files) %in% md$sample_name)){
  message("The following files do not have a matching metadata entry:")
  print(names(tx_files)[(names(tx_files) %nin% md$sample_name)])
} else {
  message("All files match a metadata entry.")
}

if (!all(md$sample_name %in% names(tx_files))){
  message("A count file was not found for the following selected samples:")
  print(md$sample_name[(md$sample_name %nin% names(tx_files))])
} else {
  message("All expected files found.")
}
```

## Sample QC filtering

Samples with a PC1 zscore > 2 (those that fall outside the green lines below) were dropped from analysis.
```{r, fig.width=6, fig.height=6}
pca_qc %>% ggplot(aes(x = PC1,
                      y = PC2,
                      color = zscore)) +
  geom_point(alpha = 0.5) +
  geom_vline(aes(xintercept = (3*sd(.data$PC1)) + mean(.data$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(3*sd(.data$PC1)) + mean(.data$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = (2*sd(.data$PC1)) + mean(.data$PC1)),
             color = "green") +
  geom_vline(aes(xintercept = -(2*sd(.data$PC1)) + mean(.data$PC1)),
             color = "green") +
  scale_color_gradient(low = "grey90",high = "red") +
  theme_cowplot()
```

```{r}
if (length(removed_outliers) > 0) {
  message("Removing the following samples:")
  md %>%
    filter(sample_name %in% removed_outliers) %>%
    kable() %>%
    kable_styling(
      bootstrap_options = c("striped", "hover", "condensed", "responsive"),
      full_width = FALSE)
}
```

### Overall correlation between samples

```{r, fig.width=12, fig.height=9}
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sample_dists,
         clustering_distance_cols = sample_dists, 
         annotation_row = annotation_info,
         show_rownames = FALSE,
         show_colnames = TRUE,
         annotation_colors = group_pal,
         angle_col = 45)
```

# Data quality

## Mean-difference plot
```{r}
plotMD(qlf)
```

## Dispersion estimates
```{r}
plotBCV(dge_qc)
```

## Quasi-likelihood dispersion

```{r}
plotQLDisp(fit)
```

## Mean-variance relationship plot

```{r}
plotMeanVar(dge_qc, show.tagwise.vars=TRUE)
```


# Differential Gene Expression {.tabset}

```{r}
print("Log fold change results:")
results <- res %>% mutate(upreg = logFC > 0, downreg = logFC < 0) %>% filter(FDR < 0.05) %>% count(upreg, downreg)

n_total <- res %>% filter(logCPM > 0) %>% nrow
n_upreg <- results %>% filter(upreg == TRUE, downreg == FALSE) %>% pull(n)
n_downreg <- results %>% filter(upreg == FALSE, downreg == TRUE) %>% pull(n)

print(str_glue("Out of {n_total} genes with > 0 counts:"))
print(str_glue("logFC > 0 (upregulated)   : {n_upreg}, {round((n_upreg/n_total)*100, 2)}%"))
print(str_glue("logFC < 0 (downregulated) : {n_downreg}, {round((n_downreg/n_total)*100, 2)}%"))
```

## 25 genes with largest negative log-fold change in SLE
```{r}
down_table %>%
  dplyr::select(-F) %>%
  dplyr::filter(logFC > 0) %>%
  arrange(desc(logFC)) %>%
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    FDR = cell_spec(FDR,
                     color = "white",
                     bold = ifelse(FDR < 0.05, T, F),
                     background = spec_color(FDR)),
    logFC = color_bar("tomato")(logFC)) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black")
```

## 25 genes with largest positive log-fold change in SLE
```{r}
up_table %>%
  select(-F) %>%
  filter(logFC > 0) %>%
  arrange(desc(logFC)) %>%
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    FDR = cell_spec(FDR,
                     color = "white",
                     bold = ifelse(FDR < 0.05, T, F),
                     background = spec_color(FDR)),
    logFC = color_bar("lightgreen")(logFC)) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black")
```

## Expression of top DE genes:
```{r, fig.width=12, fig.height=9}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
t(v$E)[,unique(c(top_up, top_down))] %>% 
    pheatmap(
      scale = 'column',
      fontsize_row = 6, 
      fontsize_col = 6, 
      annotation_row = annotation_info,
      annotation_colors = group_pal,
      angle_col = 315,
      breaks = zscore_range,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      gaps_col = class_gaps,
      border_color = NA
      )
```

## Expression of selected interferon-stimulated genes:
```{r, fig.width=10, fig.height=9}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
t(v$E)[,ISGs] %>%
    pheatmap(
      scale = 'column',
      fontsize_row = 6,
      fontsize_col = 6,
      annotation_row = annotation_info,
      annotation_colors = group_pal,
      angle_col = 315,
      breaks = zscore_range,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      gaps_col = class_gaps,
      border_color = NA)
```

## Expression of individual genes in the M1.2, M3.4, and M5.12 gene modules
```{r fig.height=9, fig.width=12}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
v$E %>%
  as_tibble(rownames = "gene") %>%
  filter(gene %in% rownames(module_ISGs)) %>%
  filter(gene != "MT1A") %>%
  column_to_rownames("gene") %>%
  t() %>%
  `[`(,rownames(module_ISGs)[(rownames(module_ISGs) != "MT1A")]) %>%
  pheatmap(
    scale = 'column',
    fontsize_row = 6, 
    fontsize_col = 6, 
    annotation_row = annotation_info,
    annotation_col = module_ISGs,
    annotation_colors = group_pal,
    cluster_cols = FALSE,
    angle_col = 315,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA)
```


# Clustering and variation of samples in reduced dimensional space {.tabset}

## PCA

### PCA, grouped by disease classification
```{r, fig.width=12, fig.height=6}
plot_grid(pca_plot +
            theme(legend.position = "none"),
          pca_plot +
            geom_mark_ellipse(aes(fill=disease_class),
                              alpha = 0.05))
```

### PCA, grouped by study group
```{r, fig.width=12, fig.height=6}
plot_grid(
  pca_res %>%
    ggplot(aes(x = PC1,
               y = PC2,
               color = study_group)) +
    geom_point(alpha = 0.3) + 
    scale_color_brewer(palette = "Set1") +
    scale_fill_brewer(palette = "Set1") +
    labs(color = "study_group") +
    theme_cowplot() +
    theme(legend.position = "none"),
  pca_res %>%
    ggplot(aes(x = PC1,
               y = PC2,
               color = study_group)) +
    geom_point(alpha = 0.3) + 
    scale_color_brewer(palette = "Set1") +
    scale_fill_brewer(palette = "Set1") +
    labs(color = "study_group") +
    theme_cowplot() + 
    geom_mark_ellipse(aes(fill=study_group),
                      alpha = 0.05)
)
```

### PCA, grouped by the plate on which the sample was run
```{r, fig.width=12, fig.height=6}
plot_grid(pca_plot2 +
            theme(legend.position = "none"),
          pca_plot2 +
            geom_mark_ellipse(aes(fill=run_id),
                              alpha = 0.05))
```

### PCA, colored by initial sample concentration
```{r, fig.width=8, fig.height=6}
pca_res %>%
  ggplot(aes(x = PC1,
             y = PC2,
             color = initial_concentration_ng_ul)) +
  geom_point(alpha = 0.3) + 
  scale_color_gradient2(low = "blue", mid = "grey90", high = "red", midpoint = 10) +
  theme_cowplot()
```

## UMAP

### UMAP, grouped by the plate on which the sample was run
```{r, fig.width=12, fig.height=6}
plot_grid(umap_plot +
            theme(legend.position = "none"),
          umap_plot +
            geom_mark_ellipse(aes(fill=disease_class), alpha = 0.05))
```

### UMAP, grouped by the plate on which the sample was run
```{r, fig.width=12, fig.height=6}
plot_grid(umap_plot2 +
            theme(legend.position = "none"),
          umap_plot2 +
            geom_mark_ellipse(aes(fill=run_id), alpha = 0.05))
```

### UMAP, grouped by study group
```{r, fig.width=12, fig.height=6}
plot_grid(
  umap_results %>%
    ggplot(aes(x = umap_1,
               y = umap_2,
               color = study_group)) +
    geom_point(alpha = 0.3) + 
    scale_color_brewer(palette = "Set1") +
    scale_fill_brewer(palette = "Set1") +
    labs(color = "study_group") +
    theme_cowplot() +
    theme(legend.position = "none"),
  umap_results %>%
    ggplot(aes(x = umap_1,
               y = umap_2,
               color = study_group)) +
    geom_point(alpha = 0.3) + 
    scale_color_brewer(palette = "Set1") +
    scale_fill_brewer(palette = "Set1") +
    labs(color = "study_group") +
    theme_cowplot() + 
    geom_mark_ellipse(aes(fill=study_group),
                      alpha = 0.05)
)
```

### UMAP, colored by initial sample concentration
```{r, fig.width=8, fig.height=6}
umap_results %>%
  ggplot(aes(x = umap_1,
             y = umap_2,
             color = initial_concentration_ng_ul)) +
  geom_point(alpha = 0.3) + 
  scale_color_gradient2(low = "blue", mid = "grey90", high = "red", midpoint = 10) +
  theme_cowplot()
```

# Module scores {.tabset}

## SLE Interferon module scores

```{r fig.height=11, fig.width=6}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
ifn_scores %>%
  column_to_rownames("sample") %>%
  pheatmap(
    scale="column",
    fontsize_row = 7,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## SLE Inflammatory module scores

```{r fig.height=11, fig.width=7}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
inflammation_scores %>%
  column_to_rownames("sample") %>%
  pheatmap(scale="column",
    fontsize_row = 7,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## Low-density granulocyte module scores

```{r fig.height=11, fig.width=6}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)
ldg_scores %>%
  column_to_rownames("sample") %>%
  pheatmap(
    scale="column",
    fontsize_row = 7,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## SLE Modules with an annotated associated pathway
```{r fig.height=9, fig.width=9}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
pheatmap(
  annotated_module_scores,
  scale = 'column',
  fontsize_row = 6, 
  fontsize_col = 6, 
  annotation_row = annotation_info,
  annotation_col = column_to_rownames(annotated_modules, var = "module"),
  annotation_colors = group_pal,
  show_rownames = FALSE,
  angle_col = 315,
  breaks = zscore_range,
  color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA)
```

## All SLE modules

```{r fig.height=9, fig.width=11}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)
column_to_rownames(.data = module_scores,
                   var = "sample") %>%
  pheatmap(
         scale = 'column',
         fontsize_row = 6, 
         fontsize_col = 6, 
         annotation_row = annotation_info,
         annotation_col = module_annot,
         annotation_colors = group_pal %>% `$<-`(type, c(.$type, Undetermined = "#000000")),
         show_rownames = FALSE,
         angle_col = 315,
           breaks = zscore_range,
  color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA)
```

# Viral trancription {.tabset}

Samples were examined for transcripts from the following viruses:
HSV-1, HSV-2, VZV, EBV, EBV-2, HCMV, HHV7, KSHV, Adenovirus 2, B19, HIV-1, HBV, HPV-2, HPV-16, JCV, and MCV-1

## Detected viral transcription and Interferon module scores

```{r fig.height=11, fig.width=8}
range_bound = 5
zscore_range <- c(-range_bound:range_bound)
ifn_scores_with_viral %>%
  column_to_rownames("sample") %>%
  pheatmap(scale="column",
           annotation_colors = group_pal,
           annotation_row = annotation_info,
             breaks = zscore_range,
  color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA,
  angle=315)
```

## Detected viral transcription and Inflammatory module scores

```{r fig.height=11, fig.width=8}
range_bound = 5
zscore_range <- c(-range_bound:range_bound)
inflammation_scores_with_viral %>%
  column_to_rownames("sample") %>%
  pheatmap(scale="column",
           annotation_colors = group_pal,
           annotation_row = annotation_info,
             breaks = zscore_range,
  color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA,
  angle=315)
```


# Sample Sanity check

To determine if there is a gross mixup of samples, examine the expression of sex-specific transcripts to ensure that only female-subject samples express "XIST" and male-subject samples "DDX3Y:

```{r}
sex_specific_genes <- tibble(chr = c(rep("Y",16), rep("X",2)),
                             symbol = c("DDX3Y","EIF1AY",
                                       "KDM5D","NLGN4Y",
                                       "PCDH11Y","PRORY",
                                       "PRY","PRY2",
                                       "RPS4Y1","RPS4Y2",
                                       "TBL1Y","TMSB4Y",
                                       "USP9Y","UTY",
                                       "VCY","ZFY",
                                       "XIST","ZFX"))

range_bound = 5
zscore_range <- c(-range_bound:range_bound)
t(v$E)[,intersect(rownames(v$E),
                  sex_specific_genes[['symbol']])] %>%
  pheatmap(
    scale='column',
    fontsize_row=6,
    fontsize_col = 6,
    annotation_row = annotation_info[,c("sex","run_id", "disease_class", "study_group")],
    annotation_colors = group_pal,
    annotation_col = column_to_rownames(sex_specific_genes, "symbol"),
    cluster_cols = FALSE,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA,
    angle=315)
```

These appear to be correct.

```{r}
v$E %>% 
  fwrite(row.names = TRUE,
         file = "filtered_voom_transformed_expression.csv")
dge_qc$samples %>%
  fwrite(row.names = TRUE,
         file = "filtered_metadata.csv")
cpm(dge_qc, log = TRUE) %>%
  as.data.frame() %>%
  fwrite(row.names = TRUE,
         file = "filtered_transcript_log_counts_per_million.csv")
res %>%
  fwrite(row.names = TRUE,
         file = "log_fold_changes.csv")

module_scores %>%
  fwrite(row.names = TRUE,
         file = "module_scores.csv")
```

```{r}
sessionInfo()
```