---
title: "MS dataset - comparison of RRMS patients by disease state"
author: "Miles Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 1500px !important;
      width: 1500px !important;
    }
    body {
      max-width: 1500px !important;
    }
    caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
```

```{r setup, include=FALSE}
options(width = 1200)
knitr::opts_chunk$set(echo = FALSE)
theme_set(theme_cowplot())
loadd(annotation_info,
  group_pal,
  module_scores_with_viral,
  ifn_modules,
  inflame_modules,
  module_ISGs,
  pca_qc,
  pca_results,
  removed_outliers,
  sample_dendrogram,
  sample_dists,
  sampleDistMatrix,
  up_table,
  down_table,
  top_up,
  top_down,
  tx_files,
  umap_results,
  viral_transcripts)
```

# Data processing of RNAseq data

## Overview of sample characteristics:

```{r fig.height=4, fig.width=10}
loadd(md_cat_data)
inspectdf::show_plot(md_cat_data, high_cardinality=1, col_palette = 3)
rm(md_cat_data)
```

```{r fig.height=4, fig.width=10}
loadd(md_num_data)
inspectdf::show_plot(md_num_data, col_palette = 3)
rm(md_num_data)
```

## Breakdown of the source of samples:

```{r}
loadd(md)
md %>%
  tabyl(disease_class) %>%
  adorn_pct_formatting(digits = 2) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, bold = TRUE, border_right = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = FALSE, width = "10em")

```

```{r}
md %>% 
  tabyl(run_id, disease_activity) %>%
  adorn_totals("col") %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, border_right = TRUE, bold = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = TRUE, width = "10em") %>%
  column_spec(4, border_right = FALSE, width = "10em")
```

```{r echo=FALSE}
message(str_glue("{length(md$sample_name)} samples selected for analysis"))
message(str_glue("{length(names(tx_files))} files found"))
if (!all(names(tx_files) %in% md$sample_name)){
  message("The following files do not have a matching metadata entry:")
  print(names(tx_files)[(names(tx_files) %nin% md$sample_name)])
} else {
  message("All files match a metadata entry.")
}

if (!all(md$sample_name %in% names(tx_files))){
  message("A count file was not found for the following selected samples:")
  print(md$sample_name[(md$sample_name %nin% names(tx_files))])
} else {
  message("All expected files found.")
}
```

## Sample QC filtering

Samples with a PC1 > 2 (the red lines) and/or PC2 zscore > 2.5 (the green lines) were dropped from analysis.
```{r, fig.width=6, fig.height=6}
pca_qc %>% ggplot(aes(x = PC1,
                      y = PC2)) +
  geom_point(shape = 21,
             colour = "black",
             fill = "grey80",
             size = 2,
             stroke = 0.5,
             alpha = 0.6) +
  geom_vline(aes(xintercept = (pc1_zscore_threshold*sd(.data$PC1)) + mean(.data$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(pc1_zscore_threshold*sd(.data$PC1)) + mean(.data$PC1)),
             color = "red") +
  geom_hline(aes(yintercept = (pc2_zscore_threshold*sd(.data$PC2)) + mean(.data$PC2)),
             color = "green") +
  geom_hline(aes(yintercept = -(pc2_zscore_threshold*sd(.data$PC2)) + mean(.data$PC2)),
             color = "green")
```

```{r}
if (length(removed_outliers) > 0) {
  filter(md, sample_name %in% removed_outliers) %>%
  kable(caption = "Removing the following samples:") %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = FALSE)
}
```

### Overall correlation between samples
```{r, fig.width=12, fig.height=9}
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sample_dists,
         clustering_distance_cols = sample_dists, 
         annotation_row = annotation_info,
         show_rownames = FALSE,
         show_colnames = TRUE,
         annotation_colors = group_pal,
         angle_col = 45)
```

# Data quality

## MA-plot
```{r echo=FALSE}
loadd(res)
for(m in seq_along(res)){
  plotMA(res[[m]], ylim = c(-6,6), main = names(res)[[m]])
}
#plotMA(res[[1]], ylim = c(-4,4))
```

## Dispersion estimates
```{r}
loadd(disp_plot)
disp_plot
rm(disp_plot)
```

# Differential Gene Expression {.tabset .tabset-fade .tabset-pills}
```{r}
message("Log fold change results:")
for (j in seq_along(res)){
  message(names(res)[[j]])
  summary(res[[j]])
}
```

## 25 genes with largest negative log-fold change in MS-like disorders
```{r echo = FALSE, results = "asis"}
loadd(down_tables)
for (i in seq_along(down_tables)){
  down_tables[[i]] %>%
    dplyr::select(-baseMean) %>%
    dplyr::filter(log2FoldChange > 0) %>%
    arrange(desc(log2FoldChange)) %>%
    mutate(
      gene = cell_spec(gene,
                       bold = TRUE),
      padj = cell_spec(padj,
                       color = "white",
                       bold = TRUE,
                       background = spec_color(padj)),
      log2FoldChange = color_bar("tomato")(log2FoldChange)) %>%
    ungroup() %>%
    rename(!!paste0("pvalue", footnote_marker_symbol(1)) := pvalue,
           !!paste0("stat", footnote_marker_symbol(2)) := stat,
           !!paste0("padj", footnote_marker_symbol(3)) := padj,
           "lfc standard error" = lfcSE) %>%
    kable(escape = FALSE,
          caption = names(down_tables)[[i]]) %>%
    kable_styling(
      bootstrap_options = c("striped", "hover", "condensed", "responsive"),
      full_width = TRUE,
      fixed_thead = TRUE) %>%
    footnote(symbol = c("Wald test p-values", "Wald statistic", "Benjamini–Hochberg adjusted value")) %>%
    column_spec(2, color = "black") %>% print()
}
```

## 25 genes with largest positive log-fold change in MS-like disorders
```{r echo = FALSE, results = "asis"}
loadd(up_tables)
for (j in seq_along(up_tables)){
  up_tables[[j]] %>%
    select(-baseMean) %>%
    filter(log2FoldChange > 0) %>%
    arrange(desc(log2FoldChange)) %>%
    mutate(
      gene = cell_spec(gene,
                       bold = TRUE),
      padj = cell_spec(padj,
                       color = "white",
                       bold = ifelse(pvalue < 0.05, T, F),
                       background = spec_color(padj)),
      log2FoldChange = color_bar("lightgreen")(log2FoldChange)) %>%
    ungroup() %>%
    rename(!!paste0("pvalue", footnote_marker_symbol(1)) := pvalue,
           !!paste0("stat", footnote_marker_symbol(2)) := stat,
           !!paste0("padj", footnote_marker_symbol(3)) := padj,
           "lfc standard error" = lfcSE) %>%
    kable(escape = FALSE,
          caption = names(up_tables)[[j]]) %>%
    kable_styling(
      bootstrap_options = c("striped", "hover", "condensed", "responsive"),
      full_width = TRUE,
      fixed_thead = TRUE) %>%
    footnote(symbol = c("Wald test p-values", "Wald statistic", "Benjamini–Hochberg adjusted value")) %>%
    column_spec(2, color = "black") %>% print()
  }
```

## Expression of top DE genes
```{r, fig.width=12, fig.height=9}
loadd(vsd_exprs, deg_means)
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
t(vsd_exprs)[,colnames(deg_means)] %>% 
    pheatmap(
      scale = 'column',
      fontsize = 9,
      fontsize_col = 6,
      show_rownames = FALSE,
      annotation_row = annotation_info,
      annotation_colors = group_pal,
      angle_col = 315,
      breaks = zscore_range,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      gaps_col = class_gaps,
      border_color = NA
      )
```

## Mean expression of disease activity-specific genes

Gene expression for each of the three disease classes - active relapse, recent recovery, and long term stable - were compared to those in controls.  Genes were considered differentially expressed if their log2 fold change was >= than 0.5 or <= -0.5 and had an adjusted p-value of < 0.05.

```{r fig.height=15, fig.width=6}
loadd(deg_class)
pheatmap(
  mat = t(deg_means),
  scale = "row",
  annotation_row = deg_class,
  annotation_colors = group_pal,
  angle_col = 45,
  fontsize_col = 9,
  fontsize_row = 6,
  color = viridis(n = 100))
```

## Volcano plot of gene expression
```{r fig.height=6, fig.width=6}

  for(k in seq_along(res)){
      volcano_data <- 
        res[[k]] %>%
        as_tibble(rownames = "gene") %>%
        mutate(sig = case_when(
          padj > 0.01 ~ "insig",
          padj < 0.01 & log2FoldChange <= -0.25 ~ "down",
          padj < 0.01 & log2FoldChange >= 0.25 ~ "up",
          padj < 0.01 & between(log2FoldChange, -1, 1) ~ "ignore",
          is.na(padj) ~ "insig")
          )
      

  top_volcano_data =
    rbind(
      top_n(filter(volcano_data, sig == "up"), n = 20, log2FoldChange),
      top_n(filter(volcano_data, sig == "down"), n = -20, log2FoldChange)
    )
    
  print(ggplot(data = volcano_data,
         mapping = aes(x = log2FoldChange,
                       y = 1/padj,
                       color = sig)) +
    geom_point(alpha = 0.25) +
    scale_y_log10() +
    theme_cowplot() +
    scale_color_manual(values = c("down" = 'red',
                                  "up" = "blue",
                                  "insig" = "black",
                                  "ignore" = 'grey')) +
    geom_text_repel(data = top_volcano_data,
                    aes(label = gene)) +
    guides(color = "none") +
    labs(caption = names(res)[[k]]))
  }
```

## Expression of selected interferon-stimulated genes
```{r, fig.width=9, fig.height=6}
loadd(ISGs)

range_bound = 4
zscore_range <- c(-range_bound:range_bound)
t(vsd_exprs)[,ISGs] %>%
    pheatmap(
      scale = 'column',
      fontsize = 6,
      show_rownames = FALSE,
      annotation_row = annotation_info,
      annotation_colors = group_pal,
      angle_col = 315,
      breaks = zscore_range,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      gaps_col = class_gaps,
      border_color = NA)

rm(ISGs)
```

## Expression of individual genes in the M1.2, M3.4, and M5.12 gene modules
```{r fig.height=6, fig.width=9}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
vsd_exprs %>%
  as_tibble(rownames = "gene") %>%
  filter(gene %in% rownames(module_ISGs)) %>%
  filter(gene != "MT1A") %>%
  column_to_rownames("gene") %>%
  t() %>%
  `[`(,rownames(module_ISGs)[(rownames(module_ISGs) != "MT1A")]) %>%
  pheatmap(
    scale = 'column',
    fontsize = 6,
    show_rownames = FALSE,
    annotation_row = annotation_info,
    annotation_col = module_ISGs,
    annotation_colors = group_pal,
    cluster_cols = FALSE,
    angle_col = 315,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA)
```

# Sample Clustering

Using gene expression distance, a sample adjacency matrix was calculated from a shared nearest neighbor graph.  This, in turn, was used to cluster the samples, with the optimal clustering resolution being determined by silhouette width at multiple resolutions.

```{r}
annotation_info %>%
  group_by(disease_class,
           cluster) %>%
  summarise(n = n()) %>%
  pivot_wider(names_from = cluster,
              values_from = n) %>%
  DT::datatable()

annotation_info %>%
  select(disease_activity, cluster) %>%
  group_by(disease_activity) %>%
  mutate(total_disease = n()) %>%
  ungroup() %>%
  group_by(disease_activity, cluster) %>%
  mutate(total_cluster_disease = n(),
         freq = total_cluster_disease/total_disease) %>%
  distinct() %>%
  ggplot(aes(
    x = disease_activity,
    y = freq,
    group = cluster,
    fill = cluster
  )) +
  geom_col() +
  labs(y = "Proportion of disease activity class in cluster") +
  scale_fill_brewer(palette = "Set1") +
  theme_cowplot()
```


# Clustering and variation of samples in reduced dimensional space {.tabset .tabset-fade .tabset-pills}

## PCA

### PCA, by disease classification
```{r, fig.width=12, fig.height=6}
plot_grid(
  ggplot(
    data = pca_results,
    aes(
      x = PC1,
      y = PC2,
      fill = disease_class,
      line = 1
    ),
    shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    theme_cowplot() +
    theme(legend.position = "none"),
  ggplot(
    data = pca_results,
    aes(
      x = PC1,
      y = PC2,
      fill = disease_class,
      line = 1
    ), shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    scale_color_brewer(palette = "Set1") +
    theme_cowplot() +
    geom_mark_ellipse(aes(fill = disease_class,
                          color = disease_class),
                      alpha = 0.05
    )
)
```

### PCA, by disease activity
```{r, fig.width=12, fig.height=6}
plot_grid(
  ggplot(data = pca_results,
         mapping = aes(x = PC1,
                       y = PC2,
                       fill = disease_activity)) +
    geom_point(shape = 21,
               colour = "black",
               size = 2,
               stroke = 0.25,
               alpha = 0.6) + 
    scale_color_paletteer_d("ggthemes::calc") +
    scale_fill_paletteer_d("ggthemes::calc") +
    theme(legend.position = "none"),
  ggplot(data = pca_results,
         mapping = aes(x = PC1,
                       y = PC2,
                       fill = disease_activity)) +
    geom_point(shape = 21,
               colour = "black",
               size = 2,
               stroke = 0.25,
               alpha = 0.6) + 
    scale_color_paletteer_d("ggthemes::calc") +
    scale_fill_paletteer_d("ggthemes::calc") +
    labs(color = "Group",
         fill = "none") +
    geom_mark_ellipse(aes(fill=disease_activity,
                          color=disease_activity),
                      alpha = 0.05)
)
```

### PCA, by plate
```{r, fig.width=12, fig.height=6}
p1 <- ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = run_id
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  labs(color = "run_id") +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv")


p2 <- ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = run_id
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  labs(color = "run_id") +
  geom_mark_ellipse(aes(
    fill = run_id,
    color = run_id
  ),
  alpha = 0.05
  ) +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv") +
  guides(color = guide_legend(ncol = 1))

legend <- get_legend(p2)

plot_grid(p1 + theme(legend.position = "none"),
  p2 + theme(legend.position = "none"),
  legend,
  nrow = 1,
  rel_widths = c(1, 1, .25)
)
```

### PCA, by initial sample concentration
```{r, fig.width=8, fig.height=6}
ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = initial_concentration_ng_ul
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.25,
    alpha = 0.6
  ) +
  scale_fill_gradient2(low = "blue",
                       mid = "grey90",
                       high = "red",
                       midpoint = 10)

```

### PCA, by Leiden cluster
```{r, fig.width=8, fig.height=6}
ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = cluster
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.25,
    alpha = 0.6
  ) +
  scale_fill_manual(values = group_pal$cluster)
```

### 3D PCA, colored by Leiden cluster
```{r}
pca_results %>%
  plot_ly(x = ~PC1,
          y = ~PC2,
          z = ~PC3,
          color = ~cluster,
          size = 2.5,
          alpha = 0.9,
          colors = group_pal$cluster,
          text = ~paste('<br>Sample: ', sample_name,
                        '<br>Classification: ', disease_class,
                        '<br>Activity: ', disease_activity,
                        '<br>Ethnicity ', race_code,
                        '<br>Cluster: ', cluster,
                        '<br>Sex: ', sex,
                        '<br>Plate: ', run_id
          )
  ) %>%
  add_markers(marker = list(size = 3,
                            line = list(color = 'rgba(0, 0, 0, .8)',
                                        width = 1.5)))
```


## UMAP

### UMAP, by disease classification
```{r, fig.width=12, fig.height=6}
plot_grid(
  ggplot(
    data = umap_results,
    aes(
      x = umap_1,
      y = umap_2,
      fill = disease_class,
      line = 1
    ),
    shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    theme_cowplot() +
    theme(legend.position = "none"),
  ggplot(
    data = umap_results,
    aes(
      x = umap_1,
      y = umap_2,
      fill = disease_class,
      line = 1
    ), shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    scale_color_brewer(palette = "Set1") +
    theme_cowplot() +
    geom_mark_ellipse(aes(fill = disease_class,
                          color = disease_class),
                      alpha = 0.05
    )
)
```

### UMAP, by disease group
```{r, fig.width=12, fig.height=6}
plot_grid(
  ggplot(data = umap_results,
         mapping = aes(x = umap_1,
                       y = umap_2,
                       fill = disease_activity)) +
    geom_point(shape = 21,
               colour = "black",
               size = 2,
               stroke = 0.25,
               alpha = 0.6) + 
    scale_color_paletteer_d("ggthemes::calc") +
    scale_fill_paletteer_d("ggthemes::calc") +
    theme(legend.position = "none"),
  ggplot(data = umap_results,
         mapping = aes(x = umap_1,
                       y = umap_2,
                       fill = disease_activity)) +
    geom_point(shape = 21,
               colour = "black",
               size = 2,
               stroke = 0.25,
               alpha = 0.6) + 
    scale_color_paletteer_d("ggthemes::calc") +
    scale_fill_paletteer_d("ggthemes::calc") +
    labs(color = "Group",
         fill = "none") +
    geom_mark_ellipse(aes(fill=disease_activity,
                          color=disease_activity),
                      alpha = 0.05)
)
```

### UMAP, by plate
```{r, fig.width=12, fig.height=6}
p1 <- ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = run_id
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  labs(color = "run_id") +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv")


p2 <- ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = run_id
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  labs(color = "run_id") +
  geom_mark_ellipse(aes(
    fill = run_id,
    color = run_id
  ),
  alpha = 0.05
  ) +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv") +
  guides(color = guide_legend(ncol = 1))

legend <- get_legend(p2)

plot_grid(p1 + theme(legend.position = "none"),
          p2 + theme(legend.position = "none"),
          legend,
          nrow = 1,
          rel_widths = c(1, 1, .25)
)

```

### UMAP, by initial sample concentration
```{r, fig.width=8, fig.height=6}
ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = initial_concentration_ng_ul
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.25,
    alpha = 0.6
  ) +
  scale_fill_gradient2(low = "blue",
                       mid = "grey90",
                       high = "red",
                       midpoint = 10)
```

### UMAP, by Leiden cluster
```{r, fig.width=7, fig.height=6}
ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = cluster
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.25,
    alpha = 0.6
  ) +
  scale_fill_manual(values = group_pal$cluster)
```

### UMAP, in 3D, colored by Leiden cluster
```{r}
umap_results %>%
  plot_ly(x = ~umap_1,
          y = ~umap_2,
          z = ~umap_3,
          color = ~cluster,
          size = 2.5,
          alpha = 0.9,
          colors = group_pal$cluster,
          text = ~paste('<br>Sample: ', sample_name,
                        '<br>Classification: ', disease_class,
                        '<br>Activity: ', disease_activity,
                        '<br>Ethnicity ', race_code,
                        '<br>Cluster: ', cluster,
                        '<br>Sex: ', sex,
                        '<br>Plate: ', run_id
          )
  ) %>%
  add_markers(marker = list(size = 3,
                            line = list(color = 'rgba(0, 0, 0, .8)',
                                        width = 1.5)))
```

# Module scores {.tabset .tabset-fade .tabset-pills}

## SLE Interferon module scores
```{r fig.height=9, fig.width=4}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name, one_of(ifn_modules)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## SLE Inflammatory module scores
```{r fig.height=9, fig.width=6}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name, one_of(inflame_modules)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## Low-density granulocyte module scores
```{r fig.height=9, fig.width=5}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(names(ldg_modules))) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## SLE Modules with an annotated associated pathway
```{r fig.height=9, fig.width=11}
loadd(annotated_modules)

range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(annotated_modules$module)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale = 'column',
    fontsize = 9,
    show_rownames = FALSE,
    annotation_row = annotation_info,
    annotation_col = column_to_rownames(annotated_modules, var = "module"),
    annotation_colors = group_pal,
    angle_col = 315,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA)

```

## All SLE modules
```{r fig.height=9, fig.width=11}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         matches("^M[[:digit:]]+")) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(scale = 'column',
           fontsize = 9,
           annotation_row = annotation_info,
           annotation_col = module_annot,
           annotation_colors = group_pal %>%
             `$<-`(type,
                   c(.$type,
                     Undetermined = "#000000")),
           show_rownames = FALSE,
           show_colnames = FALSE,
           angle_col = 315,
           breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA)
```

# Viral trancription {.tabset .tabset-fade .tabset-pills}

Samples were examined for transcripts from the following viruses:
HSV-1, HSV-2, VZV, EBV, EBV-2, HCMV, HHV7, KSHV, Adenovirus 2, B19, HIV-1, HBV, HPV-2, HPV-16, JCV, and MCV-1

## Detected viral transcription and Interferon module scores
```{r fig.height=10, fig.width=10}
range_bound = 5
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(ifn_modules,
                viral_transcripts)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA,
    angle=315)
```

## Detected viral transcription and Inflammatory module scores
```{r fig.height=10, fig.width=9}
range_bound = 5
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(inflame_modules,
                viral_transcripts)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(scale = "column",
           annotation_colors = group_pal,
           annotation_row = annotation_info,
           breaks = zscore_range,
           fontsize = 9,
           show_rownames = FALSE,
           color = viridis(n = length(zscore_range)-1, option = "D"),
           border_color = NA,
           angle=315)
```

# WGCNA

## Expression similarity dendrogram
```{r echo=FALSE, fig.height=6, fig.width=10}
loadd(wgcna_modules)
plotDendroAndColors(dendro = wgcna_modules$dendrograms[[1]],
                    colors = wgcna_modules$colors[wgcna_modules$blockGenes[[1]]],
                    groupLabels = "Module colors",
                    main = "Gene dendrogram and module colors in block 1",
                    dendroLabels = FALSE,
                    hang = 0.03,
                    addGuide = TRUE,
                    guideHang = 0.05)
```

## Expression of identified modules {.tabset .tabset-fade .tabset-pills}

### By cluster
```{r echo=FALSE, fig.height=8, fig.width=12}
loadd(wgcna_scores)

module_scores_with_viral %>%
  select(cluster,
         starts_with("ME")) %>%
  pivot_longer(-cluster,
               names_to = "module",
               values_to = "score") %>%
  mutate(cluster = as_factor(cluster)) %>%
  ggplot(aes(x = cluster,
             y = score,
             fill = cluster)) +
  geom_violin() +
  geom_jitter(size = 0.5, alpha = 0.25) +
  scale_fill_paletteer_d("ggsci::alternating_igv") +
  facet_wrap(facets = vars(module),
             scales = "free_y") +
  theme_cowplot()
```

### By disease classification
```{r echo=FALSE, fig.height=8, fig.width=12}
module_scores_with_viral %>%
  select(disease_class,
         starts_with("ME")) %>%
  pivot_longer(-disease_class,
               names_to = "module",
               values_to = "score") %>%
  mutate(disease_class = as_factor(disease_class)) %>%
  ggplot(aes(x = disease_class,
             y = score,
             fill = disease_class)) +
  geom_violin() +
  geom_jitter(size = 0.5, alpha = 0.25) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  facet_wrap(facets = vars(module),
             scales = "free_y") +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1))
```

### By disease activity
```{r echo=FALSE, fig.height=8, fig.width=12}

module_scores_with_viral %>%
  mutate(disease_activity = fct_relevel(disease_activity, "none")) %>%
  select(disease_activity,
         starts_with("ME")) %>%
  pivot_longer(-disease_activity,
               names_to = "module",
               values_to = "score") %>%
  mutate(disease_activity = as_factor(disease_activity)) %>%
  ggplot(aes(x = disease_activity,
             y = score,
             fill = disease_activity)) +
  geom_violin() +
  geom_jitter(size = 0.5, alpha = 0.25) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  facet_wrap(facets = vars(module),
             scales = "free_y") +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
```

## WGCNA eigenvalues
```{r fig.height=9, fig.width=11}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         matches("^ME")) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(scale = 'column',
           fontsize = 9,
           annotation_row = annotation_info,
           annotation_colors = group_pal %>%
             `$<-`(type,
                   c(.$type,
                     Undetermined = "#000000")),
           show_rownames = FALSE,
           show_colnames = TRUE,
           angle_col = 315,
           breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA)
```

## Gene ontology GSEA of WGCNA modules
```{r fig.height=11, fig.width=12}
loadd(MEplotting)
for(i in 1:ceiling(length(unique(MEplotting$module))/6)){
    function(i){
      print(
        ggplot(data = MEplotting,
               mapping = aes(
                 y = GeneRatio,
                 x = order,
                 color = p.adjust)
               ) +
          geom_point(stat = "identity",
                     size = 2) +
          scale_x_continuous(breaks = MEplotting$order,
                           labels = str_wrap(MEplotting$ID, width = 40),
                           expand = c(0.1,0.1)) +
          scale_color_gradient(low = "blue", high = "red") +
          labs(color = "adjusted\np value") +
          facet_wrap_paginate(~ module,
                              ncol = 2,
                              nrow = 3,
                              scales = "free",
                              page = i) +
          theme(axis.text.y = element_text(size = 6),
                axis.text.x = element_text(size = 6),
                panel.background = element_rect(fill = "white", linetype = "solid"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                panel.grid = element_line(color = "grey90"),
                axis.title.y = element_blank()) +
          coord_flip()
      )
    }
}
```


## Use of WGCNA modules in classifying clusters.

The module eigengene scores were used to train a random forest model with repeated cross validation to predict cluster identity.

The model was trained with 75% of the dataset and tested on 25%.

```{r}
loadd(wgcna_cluster_rf_cv, wgcna_cluster_test)

caret::confusionMatrix(
  data = predict(wgcna_cluster_rf_cv, wgcna_cluster_test),
  reference = as_factor(wgcna_cluster_test$cluster)
)
```

The relative importance of each eigengene in classification:

```{r}
loadd(wgcna_cluster_rf_cv_varImp)

wgcna_cluster_rf_cv_varImp
```

## Use of WGCNA modules in labeling disease class.

The module eigengene scores were used to train a random forest model with repeated cross validation to predict cluster identity.

The model was trained with 75% of the dataset and tested on 25%.

```{r}
loadd(wgcna_disease_class_rf_cv, wgcna_disease_class_test)

caret::confusionMatrix(
  data = predict(wgcna_disease_class_rf_cv, wgcna_disease_class_test),
  reference = as_factor(wgcna_disease_class_test$disease_class)
)
```

The relative importance of each eigengene in classification:

```{r}
loadd(wgcna_disease_class_rf_cv_varImp)

wgcna_disease_class_rf_cv_varImp
```

# Sample Sanity check

To determine if there is a gross mixup of samples, examine the expression of sex-specific transcripts to ensure that only female-subject samples express "XIST" and male-subject samples "DDX3Y:

```{r}
sex_specific_genes <- tibble(chr = c(rep("Y",16), rep("X",2)),
                             symbol = c("DDX3Y","EIF1AY",
                                       "KDM5D","NLGN4Y",
                                       "PCDH11Y","PRORY",
                                       "PRY","PRY2",
                                       "RPS4Y1","RPS4Y2",
                                       "TBL1Y","TMSB4Y",
                                       "USP9Y","UTY",
                                       "VCY","ZFY",
                                       "XIST","ZFX")) %>%
  filter(symbol %in% rownames(vsd_exprs))

range_bound = 5
zscore_range <- c(-range_bound:range_bound)

t(vsd_exprs)[,c(sex_specific_genes[['symbol']])] %>%
  pheatmap(
    scale='column',
    fontsize = 6,
    show_rownames = FALSE,
    annotation_row = annotation_info[,c("sex", "disease_class")],
    annotation_colors = group_pal,
    annotation_col = column_to_rownames(sex_specific_genes, "symbol"),
    cluster_cols = FALSE,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA,
    angle=315)
```

These appear to be correct.

```{r}
sessioninfo::session_info()
```