---
title: "<Project> RNAseq"
author: "Miles Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 1500px !important;
      width: 1500px !important;
    }
    body {
      max-width: 1500px !important;
    }
```

```{r setup, include=FALSE}
options(width = 1200)
knitr::opts_chunk$set(echo = FALSE)
loadd(annotation_info,
  annotated_module_scores,
  annotated_modules,
  dds_with_scores,
  ifn_scores,
  inflammation_scores,
  ISGs,
  group_pal,
  md,
  md_cat_data,
  md_dupes,
  md_num_data,
  module_ISGs,
  module_scores,
  module_tbl,
  pca_qc,
  pca_plot,
  pca_plot2,
  pca_res,
  removed_outliers,
  sample_dendrogram,
  sample_dists,
  sampleDistMatrix,
  tx_files,
  umap_plot2,
  umap_plot,
  umap_results,
  vsd_exprs,
  ea_t,
  ea_b,
  ea_m,
  aa_t,
  aa_b,
  aa_m)
```

# Import of all SLE and control RNAseq data

Overview of sample characteristics:

```{r fig.height=6, fig.width=6}
show_plot(md_cat_data, high_cardinality=1, col_palette = 3)
```

```{r}
md %>% 
  tabyl(disease_class) %>%
  adorn_totals("col") %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, border_right = TRUE, bold = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = TRUE, width = "10em") %>%
  column_spec(4, border_right = FALSE, width = "10em")
```

## Sample QC filtering

Samples with a PC1 zscore > 3 (those that fall outside the red lines below) are dropped from analysis.
```{r, fig.width=6, fig.height=6}
pca_qc %>% ggplot(aes(x = PC1,
                      y = PC2,
                      color = zscore)) +
  geom_point(alpha = 0.5) +
  geom_vline(aes(xintercept = (3*sd(pca_qc$PC1)) + mean(pca_qc$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(3*sd(pca_qc$PC1)) + mean(pca_qc$PC1)),
             color = "red") +
  scale_color_gradient(low = "grey90",high = "red") +
  theme_cowplot()
```

```{r}
if (length(removed_outliers) > 0) {
  message("Removing the following samples:")
  md %>%
    filter(sample %in% removed_outliers) %>%
    kable() %>%
    kable_styling(
      bootstrap_options = c("striped", "hover", "condensed", "responsive"),
      full_width = FALSE)
}
```

### Overall correlation between samples

```{r, fig.width=12, fig.height=9}
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sample_dists,
         clustering_distance_cols = sample_dists, 
         annotation_row = annotation_info,
         show_rownames = TRUE,
         show_colnames = TRUE,
         annotation_colors = group_pal,
         angle_col = 45)
```

### MA plot
```{r}
plotMA(dds_with_scores, ylim = c(-2,2))
```

### Dispersion estimates
```{r}
plotDispEsts(dds_with_scores)
```

### Expression of selected interferon-stimulated genes:
```{r, fig.width=10, fig.height=9}
t(vsd_exprs)[,ISGs] %>%
    pheatmap(scale = 'column',
             border_color = NA,
             fontsize_row = 6, 
             fontsize_col = 6, 
             annotation_row = annotation_info, 
             annotation_colors = group_pal,
             angle_col = 315)
```

### Expression of individual genes in the M1.2, M3.4, and M5.12 gene modules
```{r fig.height=9, fig.width=12}
vsd_exprs %>%
  as_tibble(rownames = "gene") %>%
  filter(gene %in% rownames(module_ISGs)) %>%
  filter(gene != "MT1A") %>%
  column_to_rownames("gene") %>%
  t() %>%
  pheatmap(scale = 'column',
           border_color = NA,
             fontsize_row = 6, 
             fontsize_col = 6, 
             annotation_row = annotation_info,
             annotation_col = module_ISGs,
             annotation_colors = group_pal,
             cluster_cols = FALSE,
             angle_col = 315)
```


# Clustering and variation of samples in reduced dimensional space

## PCA

```{r, fig.width=9, fig.height=4}
plot_grid(
pca_plot$data %>% 
    ggplot(aes(x = PC1,
               y = PC2,
               color = disease_class)) +
    geom_point(alpha = 0.75) + 
    scale_color_manual(values = paletteer_d("ggthemes", "calc")[1:3]) +
    scale_fill_manual(values = paletteer_d("ggthemes", "calc")[1:3]) +
    theme_cowplot(),
  pca_plot$data %>% 
    ggplot(aes(x = PC1,
               y = PC2,
               color = celltype)) +
    geom_point(alpha = 0.75) + 
    scale_color_manual(values = paletteer_d("ggthemes", "calc")[1:3]) +
    scale_fill_manual(values = paletteer_d("ggthemes", "calc")[1:3]) +
    theme_cowplot()
)
```

## UMAP

```{r, fig.width=9, fig.height=4}
plot_grid(
  umap_plot$data %>% 
    ggplot(aes(x = umap_1,
               y = umap_2,
               color = disease_class)) +
    geom_point(alpha = 0.75) + 
    scale_color_manual(values = paletteer_d("ggthemes", "calc")[1:3]) +
    scale_fill_manual(values = paletteer_d("ggthemes", "calc")[1:3]) +
    theme_cowplot(),
  umap_plot$data %>% 
    ggplot(aes(x = umap_1,
               y = umap_2,
               color = celltype)) +
    geom_point(alpha = 0.75) + 
    scale_color_manual(values = paletteer_d("ggthemes", "calc")[8:10]) +
    scale_fill_manual(values = paletteer_d("ggthemes", "calc")[8:10]) +
    theme_cowplot()
)
```

# Module scores

## Modules with an annotated associated pathway
```{r fig.height=9, fig.width=9}
pheatmap(annotated_module_scores,
         scale = 'column',
         border_color = NA,
         fontsize_row = 6, 
         fontsize_col = 6, 
         annotation_row = annotation_info,
         annotation_col = column_to_rownames(annotated_modules, var = "module"),
         annotation_colors = group_pal,
         show_rownames = FALSE,
         angle_col = 315)
```

## All modules

```{r fig.height=9, fig.width=11}
pheatmap(module_scores,
         scale = 'column',
         border_color = NA,
         fontsize_row = 6, 
         fontsize_col = 6, 
         annotation_row = annotation_info,
         annotation_col = module_annot,
         annotation_colors = group_pal %>% `$<-`(type, c(.$type, Undetermined = "#000000")),
         show_rownames = FALSE,
         angle_col = 315)
```

## Interferon module scores

```{r fig.height=11, fig.width=8}
ifn_scores %>%
  column_to_rownames("sample") %>%
  pheatmap(scale="column",
           annotation_colors = group_pal,
           annotation_row = annotation_info)
```

## Inflammatory module scores

```{r fig.height=11, fig.width=8}
inflammation_scores %>%
  column_to_rownames("sample") %>%
  pheatmap(scale="column",
           annotation_colors = group_pal,
           annotation_row = annotation_info)
```

# African American

## T cells

```{r}
current_obj = aa_t
```

### QC PCA

```{r fig.height=4, fig.width=5}
current_obj$pca_qc %>% ggplot(aes(x = PC1,
                           y = PC2,
                           color = disease_class)) +
  geom_point(alpha = 0.5) +
  geom_vline(aes(xintercept = (3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  theme_cowplot()
```

### Expression of selected immune genes

```{r fig.height=9, fig.width=7}
sample_order <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  pull(sample)
class_breakdown <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  group_by(disease_class) %>%
  summarise(count = n()) %>%
  pull(count)
class_gaps <- c(class_breakdown[[1]], class_breakdown[[1]]+class_breakdown[[2]])
current_obj$vsd[sgl$gene,sample_order] %>%
  pheatmap(scale = "row",
           cluster_rows = FALSE,
           cluster_cols = FALSE,
           annotation_col = annotation_info["disease_class"],
           annotation_row = column_to_rownames(sgl, var = "gene"),
           angle_col = 315,
           annotation_colors = group_pal,
           breaks = c(-3,-2,-1,0,1,2,3),
           color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(6),
           gaps_col = class_gaps)
```

### Differential expression testing  {.tabset}

#### Smaller gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% sgl$gene) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

#### Expanded gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% unique(c(sgl$gene, sfgl$gene))) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

## B cells

```{r}
current_obj = aa_b
```

### QC PCA

```{r fig.height=4, fig.width=5}
current_obj$pca_qc %>% ggplot(aes(x = PC1,
                           y = PC2,
                           color = disease_class)) +
  geom_point(alpha = 0.5) +
  geom_vline(aes(xintercept = (3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  theme_cowplot()
```

### Expression of selected immune genes

```{r fig.height=9, fig.width=7}
sample_order <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  pull(sample)
class_breakdown <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  group_by(disease_class) %>%
  summarise(count = n()) %>%
  pull(count)
class_gaps <- c(class_breakdown[[1]], class_breakdown[[1]]+class_breakdown[[2]])
current_obj$vsd[sgl$gene,sample_order] %>%
  pheatmap(scale = "row",
           cluster_rows = FALSE,
           cluster_cols = FALSE,
           annotation_col = annotation_info["disease_class"],
           annotation_row = column_to_rownames(sgl, var = "gene"),
           angle_col = 315,
           annotation_colors = group_pal,
           breaks = c(-3,-2,-1,0,1,2,3),
           color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(6),
           gaps_col = class_gaps)
```

### Differential expression testing  {.tabset}

#### Smaller gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% sgl$gene) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

#### Expanded gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% unique(c(sgl$gene, sfgl$gene))) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

## Monocytes

```{r}
current_obj = aa_m
```

### QC PCA

```{r fig.height=4, fig.width=5}
current_obj$pca_qc %>% ggplot(aes(x = PC1,
                           y = PC2,
                           color = disease_class)) +
  geom_point(alpha = 0.5) +
  geom_vline(aes(xintercept = (3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  theme_cowplot()
```

### Expression of selected immune genes

```{r fig.height=9, fig.width=7}
sample_order <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  pull(sample)
class_breakdown <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  group_by(disease_class) %>%
  summarise(count = n()) %>%
  pull(count)
class_gaps <- c(class_breakdown[[1]], class_breakdown[[1]]+class_breakdown[[2]])
current_obj$vsd[sgl$gene,sample_order] %>%
  pheatmap(scale = "row",
           cluster_rows = FALSE,
           cluster_cols = FALSE,
           annotation_col = annotation_info["disease_class"],
           annotation_row = column_to_rownames(sgl, var = "gene"),
           angle_col = 315,
           annotation_colors = group_pal,
           breaks = c(-3,-2,-1,0,1,2,3),
           color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(6),
           gaps_col = class_gaps)
```

### Differential expression testing  {.tabset}

#### Smaller gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% sgl$gene) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

#### Expanded gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% unique(c(sgl$gene, sfgl$gene))) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

# European American

## T cells

```{r}
current_obj = ea_t
```

### QC PCA

```{r fig.height=4, fig.width=5}
current_obj$pca_qc %>% ggplot(aes(x = PC1,
                           y = PC2,
                           color = disease_class)) +
  geom_point(alpha = 0.5) +
  geom_vline(aes(xintercept = (3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  theme_cowplot()
```

### Expression of selected immune genes

```{r fig.height=9, fig.width=7}
sample_order <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  pull(sample)
class_breakdown <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  group_by(disease_class) %>%
  summarise(count = n()) %>%
  pull(count)
class_gaps <- c(class_breakdown[[1]], class_breakdown[[1]]+class_breakdown[[2]])
current_obj$vsd[sgl$gene,sample_order] %>%
  pheatmap(scale = "row",
           cluster_rows = FALSE,
           cluster_cols = FALSE,
           annotation_col = annotation_info["disease_class"],
           annotation_row = column_to_rownames(sgl, var = "gene"),
           angle_col = 315,
           annotation_colors = group_pal,
           breaks = c(-3,-2,-1,0,1,2,3),
           color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(6),
           gaps_col = class_gaps)
```

### Differential expression testing  {.tabset}

#### Smaller gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% sgl$gene) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

#### Expanded gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% unique(c(sgl$gene, sfgl$gene))) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

## B cells

```{r}
current_obj = ea_b
```

### QC PCA

```{r fig.height=4, fig.width=5}
current_obj$pca_qc %>% ggplot(aes(x = PC1,
                           y = PC2,
                           color = disease_class)) +
  geom_point(alpha = 0.5) +
  geom_vline(aes(xintercept = (3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  theme_cowplot()
```

### Expression of selected immune genes

```{r fig.height=9, fig.width=7}
sample_order <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  pull(sample)
class_breakdown <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  group_by(disease_class) %>%
  summarise(count = n()) %>%
  pull(count)
class_gaps <- c(class_breakdown[[1]], class_breakdown[[1]]+class_breakdown[[2]])
current_obj$vsd[sgl$gene,sample_order] %>%
  pheatmap(scale = "row",
           cluster_rows = FALSE,
           cluster_cols = FALSE,
           annotation_col = annotation_info["disease_class"],
           annotation_row = column_to_rownames(sgl, var = "gene"),
           angle_col = 315,
           annotation_colors = group_pal,
           breaks = c(-3,-2,-1,0,1,2,3),
           color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(6),
           gaps_col = class_gaps)
```

### Differential expression testing  {.tabset}

#### Smaller gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% sgl$gene) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

#### Expanded gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% unique(c(sgl$gene, sfgl$gene))) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

## Monocytes

```{r}
current_obj = ea_m
```

### QC PCA

```{r fig.height=4, fig.width=5}
current_obj$pca_qc %>% ggplot(aes(x = PC1,
                           y = PC2,
                           color = disease_class)) +
  geom_point(alpha = 0.5) +
  geom_vline(aes(xintercept = (3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  geom_vline(aes(xintercept = -(3*sd(current_obj$pca_qc$PC1)) + mean(current_obj$pca_qc$PC1)),
             color = "red") +
  theme_cowplot()
```

### Expression of selected immune genes

```{r fig.height=9, fig.width=7}
sample_order <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  pull(sample)
class_breakdown <-
  colData(current_obj$dds) %>%
  as_tibble(rownames="sample") %>%
  arrange(disease_class) %>%
  group_by(disease_class) %>%
  summarise(count = n()) %>%
  pull(count)
class_gaps <- c(class_breakdown[[1]], class_breakdown[[1]]+class_breakdown[[2]])
current_obj$vsd[sgl$gene,sample_order] %>%
  pheatmap(scale = "row",
           cluster_rows = FALSE,
           cluster_cols = FALSE,
           annotation_col = annotation_info["disease_class"],
           annotation_row = column_to_rownames(sgl, var = "gene"),
           angle_col = 315,
           annotation_colors = group_pal,
           breaks = c(-3,-2,-1,0,1,2,3),
           color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(6),
           gaps_col = class_gaps)
```

### Differential expression testing  {.tabset}

#### Smaller gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% sgl$gene) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

#### Expanded gene list

```{r}
bind_cols(current_obj$res$NEG_vs_POS %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj), 
          current_obj$res$NEG_vs_SLE %>%
            as_tibble(rownames="gene") %>%
            select(gene,
                   log2FC = log2FoldChange,
                   pvalue,
                   padj)) %>%
  bind_cols(current_obj$res$POS_vs_SLE %>%
              as_tibble(rownames="gene") %>%
              select(gene,
                     log2FC = log2FoldChange,
                     pvalue,
                     padj)) %>%
  filter(gene %in% unique(c(sgl$gene, sfgl$gene))) %>%
  select(-gene1,
         -gene2) %>% 
  mutate(
    gene = cell_spec(gene,
                     bold = TRUE),
    padj = cell_spec(padj,
                     color = "white",
                     bold = TRUE,
                     background = spec_color(padj)),
    padj1 = cell_spec(padj1,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj1)),
    padj2 = cell_spec(padj2,
                      color = "white",
                      bold = TRUE,
                      background = spec_color(padj2)),
    log2FC = round(log2FC, 2),
    log2FC1 = round(log2FC1, 2),
    log2FC2 = round(log2FC2, 2),
    pvalue = cell_spec(scientific(pvalue, 2), 
                       bold = ifelse(pvalue < 0.05, T, F),
                       italic = ifelse(pvalue < 0.05, T, F),
                       color = spec_color(pvalue, option="E")),
    pvalue1 = cell_spec(scientific(pvalue1, 2), 
                        bold = ifelse(pvalue1 < 0.05, T, F),
                        italic = ifelse(pvalue1 < 0.05, T, F),
                        color = spec_color(pvalue1, option="E")),
    pvalue2 = cell_spec(scientific(pvalue2, 2), 
                        bold = ifelse(pvalue2 < 0.05, T, F),
                        italic = ifelse(pvalue2 < 0.05, T, F),
                        color = spec_color(pvalue2, option="E"))) %>%
  ungroup() %>%
  kable(escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = TRUE,
    fixed_thead = TRUE) %>%
  column_spec(2, color = "black") %>%
  column_spec(5, color = "black") %>%
  column_spec(8, color = "black") %>%
  add_header_above(c("",
                     "POS/NEG" = 3,
                     "SLE/NEG" = 3,
                     "SLE/POS" = 3))
```

```{r}
sessionInfo()
```