---
title: "CLE RNAseq analysis QC"
author: "Miles Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document: default
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 1500px !important;
      width: 1500px !important;
    }
    body {
      max-width: 1500px !important;
    }
    caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
```

```{r setup, include=FALSE}
source("/home/rstudio/workspace/analysis/rnaseq/cle/R/packages.R")
load(file="/home/rstudio/workspace/analysis/rnaseq/cle/external_data/gencode_v32_virus_tx2gene_v1.2.RData") #generated from rtracklayer::readGFF()

load(file="/home/rstudio/workspace/analysis/rnaseq/cle/external_data/banchereau-modules.RData")

load(file="/home/rstudio/workspace/analysis/rnaseq/cle/external_data/kegerreis-ldg-modules.RData")

metasignature_module <- list(mg = c("IFI44L", "EPSTI1", "HERC5",
                        "IFI44", "IFI27", "RSAD2",
                        "ISG15", "OASL", "IFIT3",
                        "CMPK2", "IFIT1", "USP18",
                        "SIGLEC1", "MX2", "MX1",
                        "LY6E", "PLSCR1", "OAS2",
                        "SPATS2L", "OAS1", "OAS3",
                        "PARP12", "SCO2", "IFIH1",
                        "IFITM3", "IFITM1", "DDX60",
                        "TRIM22", "RTP4", "SAMD9L",
                        "XAF1", "IFI35", "TNFAIP6",
                        "MT2A", "LAP3", "HERC6",
                        "FBXO6", "TDRD7", "IFIT5",
                        "PHF11", "IFIT2", "TAP1",
                        "TOR1B", "TNFSF13B", "ELANE",
                        "IRF7", "DHX58", "ZBP1",
                        "TYMP", "GRN", "LAMP3",
                        "IFI6", "NTNG2", "SAT1",
                        "SERPING1", "DEFA4", "DDX58",
                        "SAMD9", "PARP9", "MT1E",
                        "IRF9", "TCN2", "MT1HL1",
                        "HSH2D", "ISG20", "ZCCHC2",
                        "SP100", "LMO2", "GBP1",
                        "IFITM2", "LGALS3BP", "REM2",
                        "TNFSF10", "HESX1", "MT1A",
                        "CCR1", "CEACAM1", "MYD88",
                        "BST2", "LHFPL2", "FFAR2",
                        "MT1F", "RPL10A", "CD1C",
                        "VSIG1", "GPR183", "DSC1",
                        "KLRB1", "FBL", "EIF3E",
                        "ABCB1", "NAP1L3", "EIF3L"))

cytokine_names <- c("baff" = "BAFF",
                    "blc_cxcl13" = "BCL-CXCL13",
                    "gm_csf" = "GM-CSF",
                    "gro_alpha" = "GROalpha",
                    "ifn_alpha" = "IFNalpha",
                    "ifn_gamma" = "IFNgamma",
                    "il1alpha" = "IL1alpha",
                    "il1beta" = "IL1beta",
                    "il_1ra" = "IL1RA",
                    "il_10" = "IL10",
                    "il_12p70" = "IL12p70",
                    "il_13" = "IL13",
                    "il_15" = "IL15",
                    "il_17a" = "IL17a",
                    "il_18" = "IL18",
                    "il_2" = "IL2",
                    "il_21" = "IL21",
                    "il22" = "IL22",
                    "il_23" = "IL23",
                    "il_27" = "IL27",
                    "il_31" = "IL31",
                    "il_4" = "IL4",
                    "il5" = "IL5",
                    "il_7" = "IL7",
                    "il_8" = "IL8",
                    "il_9" = "IL9",
                    "ip_10" = "IP10",
                    "mcp1" = "MCP1",
                    "mip1alpha" = "MIP1alpha",
                    "mip_1beta" = "MIP1beta",
                    "rantes" = "RANTES",
                    "s_cd40l" = "sCD40L",
                    "scf" = "SCF",
                    "sdf_1alpha" = "SDF1alpha",
                    "s_icam_1" = "sICAM1",
                    "tnf_alpha" = "TNFalpha",
                    "tnf_beta" = "TNFbeta")

### Setup project variables
projects_to_include = "BChong2019.1"
projects_to_exclude = c("ALE06", "Xencor")
disease_classes_to_include = NULL
disease_classes_to_exclude = NULL
samples_to_remove = c("UTSW-BC02","UTSW-BC29","UTSW-BC48", "UTSW_BC38")
study_design = ~ disease_class
comparison_grouping_variable = "disease_class"
control_group = "Control"
experimental_group = "DLE"
batch_variable = "run_id"
num_sva = 3

initial_concentration_threshold = 1.5
pc1_zscore_threshold = 2
pc2_zscore_threshold = 2.5

### Setup file locations
seq_file_directory = "/home/rstudio/workspace/datasets/rnaseq/novaseq"
metadata_file = "/home/rstudio/workspace/analysis/rnaseq/cle/metadata/NovaSeq_Sample_List.xlsx"
clinical_file = "/home/rstudio/workspace/analysis/rnaseq/cle/metadata/study_data.csv"
sledai_bpx_file = "/home/rstudio/workspace/analysis/rnaseq/cle/metadata/SLEDAI_BPX.xlsx"
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning=FALSE)
theme_set(theme_pubr())
```


# Figure 1: Correlation of gene expression modules and soluble mediators.
```{r }
loadd(
  cytokine_vs_modules,
  cytokine_vs_modules_correlation,
  cytokine_vs_modules_pivot,
  cytokine_vs_modules_pivot_correlation,
  module_score_correlation
  )

rownames(cytokine_vs_modules_correlation) <- NULL
cytokine_vs_modules_correlation <- column_to_rownames(cytokine_vs_modules_correlation, var = "rowname") %>% as.matrix()
```

## Ly's previous version with updated data.  Correlation among annotated modules and cytokines, only significant correlations (*p* < 0.05).
```{r figure 1 Lys version, fig.height=7.5, fig.width=7.5}
cytokine_vs_modules_correlation %>%
  cor_reorder() %>%
  cor_plot(
    method = "circle",
    type = "upper",
    palette = colorRampPalette(c("purple4","white","yellow4"))(200),
    tl.cex = 0.37,
    insignificant = "blank")
```

## Correlation between annotated modules and cytokines, including non-significant correlations.
```{r cytokines vs all annotated modules, fig.height=9, fig.width=9}
cytokine_vs_modules_pivot_correlation %>%
  mutate(cytokine = 
           cytokine %>%
           str_replace_all(pattern = "alpha", replacement = "&alpha;") %>%
           str_replace_all(pattern = "beta", replacement = "&beta;"),
         module = 
           module %>%
           str_replace_all(pattern = "_", replacement = " - ")) %>%
  ggplot(
    aes(
      x = module,
      y = cytokine,
      fill = cor)
    ) +
  geom_tile() +
  scale_fill_gradient2(low = "purple4", mid = "white", high = "yellow4") +
  labs(fill = "correlation") +
  theme(
    axis.text.x =
      element_text(
        size = 9,
        angle = 45,
        hjust = 1,
        vjust = 1
        ),
    axis.text.y =
      element_markdown(
        size = 9
      )
    )
```

## Correlation between annotated modules and cytokines, only significant correlations (*p* < 0.05).

```{r significant cytokine vs annotated modules, fig.height=6, fig.width=6}
cytokine_vs_modules_pivot_correlation %>%
  filter(p < 0.05) %>%
  mutate(cytokine = 
           cytokine %>%
           str_replace_all(pattern = "alpha", replacement = "&alpha;") %>%
           str_replace_all(pattern = "beta", replacement = "&beta;"),
         module = 
           module %>%
           str_replace_all(pattern = "_", replacement = " - ")) %>%
  ggplot(
    aes(
      x = module,
      y = cytokine,
      fill = cor)
    ) +
  geom_tile() +
  scale_fill_gradient2(low = "purple4", mid = "white", high = "yellow4") +
#  scale_fill_paletteer_c("scico::vik") +
  labs(fill = "correlation") +
  theme(
    axis.text.x =
      element_text(
        size = 9,
        angle = 45,
        hjust = 1,
        vjust = 1
        ),
    axis.text.y =
      element_markdown(
        size = 9
      )
    )
```

## Correlation between annotated modules.  Only significant (*p* < 0.05) correlations shown.

```{r module correlation, fig.height=9, fig.width=9}
module_score_correlation %>%
  cor_reorder() %>%
  cor_plot(
    method = "circle",
    type = "upper",
    palette = colorRampPalette(c("purple4","white","yellow4"))(200),
    tl.cex = 0.6,
    insignificant = "blank")
```

```{r}
loadd(
  sample_cluster_info,
  group_pal,
  annotated_modules,
  annotated_mod_list,
  module_scores_with_viral,
  annotation_info,
  group_pal
  )

k_clusters <- sample_cluster_info$kmeans_res
```

```{r generate cluster plot, fig.height=5, fig.width=6, message=FALSE, warning=FALSE}
rf_cluster_pca <-
  fviz_cluster(
    object = k_clusters,
    data = sample_cluster_info$rf_distance,
    ellipse.type = "convex",
    geom = "point",
    pointsize = 1,
    ggtheme = theme_pubr(),
    main = ""
  ) +
  scale_fill_manual(values = group_pal$cluster, guide = guide_legend(nrow = 2, title.position = "top")) +
  scale_color_manual(values = group_pal$cluster, guide = guide_legend(nrow = 2, title.position = "top")) +
  labs(fill = "cluster")
lgd <- get_legend(rf_cluster_pca)
```

```{r SLE Modules with an annotated associated pathway, fig.height=9, fig.width=11}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_order <- c(
  "M1.1",
  
  "M2.3",
  "M3.1",
  "M6.18",
  
  "M1.2",
  "M3.4",
  "M5.12",
  
  "M3.2",
  "M4.2",
  "M4.6",
  "M4.13",
  "M5.1",
  "M5.7",
  "M7.1",
  
  "M3.6",
  "M8.46",
  "M4.1",
  "M4.15",
  "M4.10",
  "M4.11",
  
  "M4.14",
  "M5.15",
  "LDG1.1",
  "LDG2.1",
  "M8.83",
  
  "M6.6",
  "M2.2",
  "M3.3",
  "M3.5",
  "M4.7",
  "M6.11",
  "M6.16",
  "M9.42",
  "M6.13",
  "M4.3",
  "M4.5",
  "M5.9",
  "M5.6",
  "M5.10",
  "M6.2",
  "M6.12"
)

module_gaps <- c(1,4,7,14,20,25)

sample_cluster_order_info <-
  annotation_info %>%
  as_tibble(rownames = "subject") %>%
  arrange(cluster)

sample_cluster_order_count <-
  sample_cluster_order_info %>%
  count(cluster) %>%
  pull(n) %>%
  accumulate(sum)

annotated_modules <-
  bind_rows(
    annotated_modules,
    tibble(
      module = c("LDG1.1","LDG2.1"),
      type = c("Low density granulocytes", "Low density granulocytes")
      )
    ) %>%
  mutate(type = as_factor(type))

group_pal$type <-
  set_names(
    x = as.character(
      paletteer::paletteer_d(
        palette = "ggsci::category20_d3",
        n = length(levels(annotated_modules$type))
        )
      ),
    nm = levels(annotated_modules$type)
    )
```

```{r annotated module cluster sort heatmap, fig.width=12, fig.height=9}
cluster_module_scores <- 
  module_scores_with_viral %>%
    arrange(cluster) %>%
    select(
      sample_name,
      one_of(module_order)
    ) %>%
    pivot_longer(
      cols = one_of(module_order),
      names_to = "module",
      values_to = "score"
    ) %>%
    mutate(
      module =
        recode(
          module,
          !!!annotated_mod_list
        )
    ) %>%
    pivot_wider(
      names_from = "module",
      values_from = "score"
    ) %>%
    column_to_rownames("sample_name") %>%
    t() %>%
    pheatmap(
      scale="row",
      fontsize = 9,
      show_rownames = TRUE,
      show_colnames = FALSE,
      annotation_colors = group_pal,
      annotation_legend = TRUE,
      #annotation_row = column_to_rownames(mutate(annotated_modules, module = recode(module, !!!annotated_mod_list)), var = "module"),
      annotation_col = arrange(annotation_info[,"cluster",drop=FALSE], cluster),
      breaks = zscore_range,
      cluster_rows = FALSE,
      cluster_cols = FALSE,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      angle = 315,
      gaps_col = sample_cluster_order_count,
      gaps_row = module_gaps,
      border_color = NA,
      silent = TRUE)
```

# Figure 2: Clustering of CLE patients based on gene expression module scores.

```{r cluster pca with heatmap, fig.height=5, fig.width=10}
plot_grid(
  plot_grid(lgd,
            rf_cluster_pca+theme(legend.position = "none", axis.title = element_blank()),
            ncol = 1,
            rel_heights = c(0.2, 0.8)),
  as.ggplot(as.grob(cluster_module_scores)),
  labels = c("A", "B"),
  label_y = c(1.1, 1.2),
  nrow = 1,
  rel_widths = c(1, 4)
  )
```

```{r}
loadd(study_md)
```

# Figure 3
```{r fig.height=8, fig.width=8, message=FALSE, warning=FALSE}
clasi_a <-
  study_md %>%
  select(
    clasi_activity,
    cluster
    ) %>%
  ggerrorplot(
      x = "cluster",
      y = "clasi_activity",
      add = "median_iqr",
      error.plot = "errorbar"
    ) +
  geom_dotplot(binaxis='y', stackdir='center', aes(fill = cluster)) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  stat_mean() +
  labs(
    x = "Cluster",
    y = "CLASI-A Score"
    ) +
  lims(y = c(0,25))

clasi_d <-
  study_md %>%
  select(
    clasi_damage,
    cluster
    ) %>%
  ggerrorplot(
      x = "cluster",
      y = "clasi_damage",
      add = "median_iqr",
      error.plot = "errorbar"
    ) +
  geom_dotplot(binaxis='y', stackdir='center', aes(fill = cluster)) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  stat_mean() +
  labs(
    x = "Cluster",
    y = "CLASI-D Score"
    ) +
  lims(y = c(0,25))

acr_sle_criteria <-
  study_md %>%
  select(
    cluster,
    discoid_rash,
    malar_rash,
    ana,
    photo_sensitivity,
    blood_disorder,
    oral_ulcers,
    immune_disorder
    ) %>%
  pivot_longer(-cluster, names_to = "criteria") %>%
  group_by(
    criteria,
    cluster
    ) %>%
  filter(value == TRUE) %>%
  summarise(
    count = n(),
    .groups = "drop"
    ) %>%
  group_by(criteria) %>%
  mutate(
    total = sum(count),
    fraction = count/total,
    criteria =
      recode(criteria,
             "discoid_rash" = "Discoid Rash",
             "malar_rash" = "Malar Rash",
             "ana" = "ANA",
             "photo_sensitivity" = "Photosensitivity",
             "blood_disorder" = "Hematologic Disorder",
             "oral_ulcers" = "Oral Ulcers",
             "immune_disorder" = "Immunologic Disorder"
             ) %>%
      paste0(., " (n=", total, ")")
    ) %>%
  arrange(total) %>%
  ggplot(
    aes(
      x = fct_reorder(criteria, total),
      y = fraction * 100,
      fill = cluster)
    ) +
  geom_col() +
  coord_flip() +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  labs(
    x = "ACR SLE Criteria",
    y = "Percentage of Patients"
    )

plot_grid(
  plot_grid(clasi_a, clasi_d, labels = c("A", "B"), label_size = 12, ncol = 2),
  acr_sle_criteria, labels = c("","C"), label_size = 12, nrow = 2)
```
```{r}
loadd(annotated_module_scores,
      ldg_module_scores,
      clusters)

radar_module_scores <-
  left_join(
  as_tibble(
    annotated_module_scores,
    rownames = "sample_name"
    ),
  ldg_module_scores
  ) %>%
  left_join(clusters) %>%
  set_names(
    nm = recode(
      names(.),
      !!!annotated_mod_list
      )
    ) %>%
  select(-sample_name) %>%
  pivot_longer(
    -cluster,
    names_to = "module",
    values_to = "score"
  ) %>%
  group_by(cluster,module) %>%
  summarise(avg = mean(score), .groups = "drop") %>%
  group_by(module) %>%
  mutate(scaled_avg = rescale(avg)) %>%
  select(-avg) %>%
  pivot_wider(
    names_from = "module",
    values_from = "scaled_avg"
  ) %>%
  select(
    cluster,
    `M5.15 - Neutrophils`,
    LDG1.1,
    `M4.1 - T cells`,
    `M4.15 - T cells`,
    `M6.13 - Cell Death`,
    `M6.6 - Apoptosis / Survival`,
    `M7.1 - Inflammation`,
    `M5.7 - Inflammation`,
    `M5.1 - Inflammation`,
    `M4.13 - Inflammation`,
    `M4.6 - Inflammation`,
    `M4.2 - Inflammation`,
    `M3.2 - Inflammation`,
  )
  
type_tbl <- tibble(
  module = c(
    "M5.15 - Neutrophils",
    "LDG1.1",
    "M4.1 - T cells",
    "M4.15 - T cells",
    "M6.13 - Cell Death",
    "M6.6 - Apoptosis / Survival",
    "M7.1 - Inflammation",
    "M5.7 - Inflammation",
    "M5.1 - Inflammation",
    "M4.13 - Inflammation",
    "M4.6 - Inflammation",
    "M4.2 - Inflammation",
    "M3.2 - Inflammation"),
  grouping = c(
    1,
    2,
    3,
    3,
    4,
    5,
    6,
    6,
    6,
    6,
    6,
    6,
    6
  )
)

module_groupings <-
  type_tbl %>%
  mutate(
    total = n()
    ) %>%
  add_count(grouping) %>%
  group_by(grouping) %>%
  summarise(pct = n/total) %>%
  distinct() %>%
  mutate(
    grouping = as_factor(grouping) %>% fct_rev(),
    pos = "1"
    )
```

# Figure 4: Molecular profiles of five CLE patient clusters.
```{r fig.height=10, fig.width=10}
circle_label <-
  ggplot(
  data = module_groupings,
  aes(
    x = pos,
    y = pct,
    fill = grouping
    )
  ) +
  geom_col(position = "stack", width = 0.05) +
  coord_polar(theta = "y",start = pi/1.18) +
  scale_x_discrete(limits = c("","1")) +
  guides(fill = "none") +
  theme_void() +
  scale_fill_paletteer_d("awtools::mpalette")

module_radar <- 
  ggradar(
    radar_module_scores,
    legend.text.size = 11,
    label.centre.y = FALSE,
    axis.label.size = 4,
    grid.label.size = 0,
    background.circle.colour = "white",
    legend.position = "right",
    group.point.size = 0,
    group.colours = paletteer_d("ggsci::uniform_startrek"),
    legend.title = "Cluster",
  ) +
    theme(legend.title = element_text(size = 12, hjust = 0.25))

ggdraw() +
  draw_plot(module_radar) +
  draw_plot(circle_label, scale = 0.7, x = -0.055, y = 0.0025)
```

# Supplemental Material
# Figure S1. Interferon, protein synthesis, and mitochondrial module scores of CLE patients and healthy controls.
```{r}
s1_module_scores <-
  left_join(
    as_tibble(
      annotated_module_scores,
      rownames = "sample_name"
      ),
    select(
      study_md,
      sample_name,
      disease_class
      )
    ) %>%
  select(
    disease_class,
    M1.2,
    M3.4,
    M5.12,
    M5.9,
    M6.2
    ) %>%
  set_names(
    nm = recode(
      names(.),
      !!!annotated_mod_list
      )
    ) %>%
  mutate(
    disease_class =
      fct_recode(
        .f = disease_class,
        CLE = "SCLE", CLE = "DLE", CLE = "TLE", CLE = "LP"
        )
    ) %>%
  pivot_longer(
    -disease_class,
    names_to = "module",
    values_to = "score"
  )

s1_module_scores_stats <-
  s1_module_scores %>%
  group_by(module) %>%
  wilcox_test(score ~ disease_class, ref.group = "Control", p.adjust.method = "BH") %>%
  add_significance() %>%
  add_xy_position()
```

```{r fig.height=9, fig.width=8}
ifn_error_plots <- map(unique(s1_module_scores[["module"]]), function(i){
    ggerrorplot(
      data = filter(s1_module_scores, module == i),
      x = "disease_class",
      y = "score",
      color = "disease_class",
      palette = "aaas",
      xlab = "",
      ylab = "Module Score",
      add = c("jitter", "median_iqr"),
      scales = "free",
      error.plot = "errorbar",
      ggtheme = theme_pubr(),
      width = 1,
    ) +
    guides(color = "none") +
    stat_pvalue_manual(
      data = filter(s1_module_scores_stats, module == i),
      hide.ns = TRUE
      ) +
    theme(strip.background = element_blank(),
          strip.text = element_text(size = 12, face = "bold"),
          )
})  
```

```{r fig.height=6, fig.width=8}
ggarrange(plotlist = ifn_error_plots, labels = "AUTO", align = "hv")
```

```{r}
ab_disease_fractions <-
  study_md %>%
  select(
    disease_class,
    ANA = ana_flag,
    dsDNA = ds_dna_flag,
    Chromatin = chromatin_flag,
    `SS-A` = ssa_ro_composite_flag,
    `SS-B` = ssb_la_flag,
    Sm = sm_flag,
    `Sm-RNP` = sm_rnp_flag,
    RNP = rnp_composite_flag,
    `Scl-70` = scl_70_flag,
    `Centromere B` = centromere_b_flag,
    `Ribosomal-P` = ribosomal_p_flag,
    `Jo-1` = jo_1_flag) %>%
  mutate(across(where(is.logical), as.numeric)) %>%
  pivot_longer(
    -disease_class,
    names_to = "antibody"
  ) %>%
  mutate(
    value = replace_na(data = value, replace = 0),
    disease_class =
      fct_recode(
        .f = disease_class,
        CLE = "SCLE", CLE = "DLE", CLE = "TLE", CLE = "LP"
      )
  ) %>%
  group_by(disease_class, antibody) %>%
  mutate(ab_count = sum(value)) %>%
  add_count(disease_class, name = "total") %>%
  group_by(disease_class, antibody) %>%
  summarise(
    fraction_pos = ab_count/total,
    ab_pos = ab_count,
    ab_neg = total - ab_count,
    .groups = "drop") %>%
  distinct()

ab_disease_fractions_formatted_tbl <- 
  ab_disease_fractions %>%
  mutate(
    desired = 
      str_glue("{ab_pos} ({format(x = fraction_pos*100, digits = 2, drop0trailing = TRUE)}%)") %>%
      as.character()
  ) %>%
  select(
    disease_class,
    antibody,
    desired
  ) %>%
  distinct() %>%
  pivot_wider(
    names_from = "disease_class",
    values_from = "desired"
  ) %>%
  relocate(Control, .after = CLE)

antibody_disease_pvals <- 
  map_dfr(unique(ab_disease_fractions$antibody), function(i){
    ab_disease_fractions %>%
      filter(antibody == i) %>%
      select(
        -antibody,
        -fraction_pos
      ) %>%
      column_to_rownames("disease_class") %>%
      fisher_test() %>%
      mutate(antibody = i) %>%
      relocate(antibody)
  }) %>%
  mutate(
    `P-value` =
      case_when(
        p == 1 ~ ">0.99",
        TRUE ~ format(x = p, digits = 2, drop0trailing = TRUE)
      )
  ) %>%
  select(
    antibody,
    `P-value`
  )

antibody_disease_tbl <- 
  left_join(
    ab_disease_fractions_formatted_tbl,
    antibody_disease_pvals
  ) %>%
  rename(`Autoantibody, n (%)` = antibody)
```

# Table S1.  Number of CLE patients (N=67) and controls (N=10) with positive serum autoantibodies
```{r fig.height=9, fig.width=6}
qflextable(data = antibody_disease_tbl) %>%
  bold(part = "header") %>%
  align(align = "center", part = "header") %>%
  align(align = "center", part = "body", j = 2:4) %>%
  align(align = "left", part = "body", j = 1) %>%
  flextable::footnote(
    i = 1,
    j = 4,
    part = "header",
    value = as_paragraph("Autoantibody positivity was compared between control and CLE groups via Fisher’s Exact test."),
    ref_symbols = c("a"))
```

```{r}
loadd(dds_with_scores)
```

```{r}
cytokines <-
  colData(dds_with_scores) %>%
  as_tibble(rownames = "sample_name") %>%
  left_join(select(study_md, sample_name, cluster)) %>%
  select(
    sample_name,
    disease_class,
    cluster,
    one_of(cytokine_names),
    `BCL/CXCL13` = `BCL.CXCL13`,
    `GM-CSF` = `GM.CSF`
    )

cytokine_tbl <-
  cytokines %>%
  mutate(
    disease_class =
      fct_recode(
        .f = disease_class,
        CLE = "SCLE", CLE = "DLE", CLE = "TLE", CLE = "LP"
        )
    ) %>%
  pivot_longer(
    cols = where(is.numeric),
    names_to = "cytokine"
  ) %>%
  mutate(
    value = replace_na(data = value, replace = 0),
  ) %>%
  filter(sample_name %in% study_md$sample_name)

cytokine_cluster_tbl_summary <-
  cytokine_tbl %>%
  group_by(cluster, cytokine) %>%
  summarise(
    avg = mean(value),
    iqr = IQR(value),
    n = n()
    ) %>%
  distinct()

cytokine_cluster_tbl_formatted_tbl <-
  cytokine_cluster_tbl_summary %>%
  mutate(
    desired =
      str_glue("{round(avg, 1)}\n({round(avg-iqr, 1)}-{round(avg+iqr, digits = 1)})") %>%
      as.character(),
    cluster = str_glue("Cluster {cluster}\n(N={n})")
  ) %>%
  select(
    cluster,
    cytokine,
    desired
  ) %>%
  distinct() %>%
  pivot_wider(
    names_from = "cluster",
    values_from = "desired"
  )

cytokine_cluster_pvals <-
  map_dfr(unique(cytokine_tbl$cytokine), function(i){
    cytokine_tbl %>%
      filter(cytokine == i) %>%
      kruskal_test(value ~ cluster) %>%
      mutate(
        cytokine = i,
        p = round(p, 2))
  }) %>%
  select(
    cytokine,
    `P-value` = p
    )

cytokine_cluster_compiled_tbl <-
  left_join(
    cytokine_cluster_tbl_formatted_tbl,
    cytokine_cluster_pvals
  ) %>%
  rename(`Soluble Mediator median (IQR)` = cytokine)

cytokine_disease_tbl_summary <-
  cytokine_tbl %>%
  group_by(disease_class, cytokine) %>%
  summarise(
    avg = mean(value),
    iqr = IQR(value),
    n = n()
  ) %>%
  distinct()

cytokine_disease_tbl_formatted_tbl <-
  cytokine_disease_tbl_summary %>%
  mutate(
    desired =
      str_glue("{round(avg, 1)}\n({round(avg-iqr, 1)}-{round(avg+iqr, digits = 1)})") %>%
      as.character()
  ) %>%
  select(
    disease_class,
    cytokine,
    desired
  ) %>%
  distinct() %>%
  pivot_wider(
    names_from = "disease_class",
    values_from = "desired"
  )

cytokine_disease_pvals <-
  map_dfr(unique(cytokine_tbl$cytokine), function(i){
    cytokine_tbl %>%
      filter(cytokine == i) %>%
      wilcox_test(value ~ disease_class, p.adjust.method = "BH") %>%
      mutate(
        cytokine = i,
        p = round(p, 2))
  }) %>%
  select(
    cytokine,
    `P-value` = p
  )

cytokine_disease_compiled_tbl <-
  left_join(
    relocate(cytokine_disease_tbl_formatted_tbl, Control, .after = CLE),
    cytokine_disease_pvals
  ) %>%
  rename(`Soluble Mediator median (IQR)` = cytokine)

```

# Table S2. Serum levels of soluble mediators in CLE patients (N=67) and controls (N=10).
```{r fig.height=9, fig.width=6}
qflextable(data = cytokine_disease_compiled_tbl) %>%
  bold(part = "header") %>%
  align(align = "center", part = "header") %>%
  align(align = "center", part = "body", j = 2:4) %>%
  flextable::footnote(
    i = 1,
    j = 4,
    part = "header",
    value = as_paragraph("Serum soluble mediators were compared between healthy and CLE patient groups using Wilcox rank sum test, Benjamini & Hochberg adjusted."),
    ref_symbols = c("a"))
```

```{r}
ab_cluster_fractions <-
  study_md %>%
  select(
    cluster,
    ANA = ana_flag,
    dsDNA = ds_dna_flag,
    Chromatin = chromatin_flag,
    `SS-A` = ssa_ro_composite_flag,
    `SS-B` = ssb_la_flag,
    Sm = sm_flag,
    `Sm-RNP` = sm_rnp_flag,
    RNP = rnp_composite_flag,
    `Scl-70` = scl_70_flag,
    `Centromere B` = centromere_b_flag,
    `Ribosomal-P` = ribosomal_p_flag,
    `Jo-1` = jo_1_flag) %>%
  mutate(across(where(is.logical), as.numeric)) %>%
  pivot_longer(
    -cluster,
    names_to = "antibody"
  ) %>%
  mutate(value = replace_na(value, 0)) %>%
  group_by(cluster, antibody) %>%
  mutate(ab_count = sum(value)) %>%
  add_count(cluster, name = "total") %>%
  group_by(cluster, antibody) %>%
  summarise(
    fraction_pos = ab_count/total,
    ab_pos = ab_count,
    ab_neg = total - ab_count,
    total = total,
    .groups = "drop") %>%
  distinct()

ab_cluster_fractions_formatted_tbl <- 
  ab_cluster_fractions %>%
  mutate(
    desired = 
      str_glue("{ab_pos} ({format(x = fraction_pos*100, digits = 2, drop0trailing = TRUE)}%)") %>%
      as.character(),
    cluster = 
      str_glue("Cluster {cluster}\n(N={total})")
    ) %>%
  select(
    cluster,
    antibody,
    desired,
  ) %>%
  distinct() %>%
  pivot_wider(
    names_from = "cluster",
    values_from = "desired"
  )

antibody_cluster_pvals <- 
  map_dfr(unique(ab_cluster_fractions$antibody), function(i){
    ab_cluster_fractions %>%
      filter(antibody == i) %>%
      select(
        -antibody,
        -fraction_pos
      ) %>%
      column_to_rownames("cluster") %>%
      fisher_test() %>%
      mutate(antibody = i) %>%
      relocate(antibody)
  }) %>%
  mutate(
    `P-value` =
      case_when(
        p == 1 ~ ">0.99",
        TRUE ~ format(x = p, digits = 2, drop0trailing = TRUE)
      )
  ) %>%
  select(
    antibody,
    `P-value`
  )

antibody_cluster_tbl <- 
  left_join(
    ab_cluster_fractions_formatted_tbl,
    antibody_cluster_pvals
  ) %>%
  rename(`Autoantibody, n (%)` = antibody)

```

# Table S3. CLE patients with positive serum autoantibodies in CLE patient clusters.

```{r fig.height=9, fig.width=6}
qflextable(data = antibody_cluster_tbl) %>%
  bold(part = "header") %>%
  align(align = "center", part = "header") %>%
  align(align = "center", part = "body", j = 2:7) %>%
  align(align = "left", part = "body", j = 1) %>%
  flextable::footnote(
    i = 1,
    j = 7,
    part = "header",
    value = as_paragraph("Autoantibody positivity was compared between clusters via Fisher’s Exact test."),
    ref_symbols = c("a"))
```


# Table S4. Soluble mediator levels of CLE patient clusters
```{r fig.height=9, fig.width=6}
qflextable(data = cytokine_cluster_compiled_tbl) %>%
  bold(part = "header") %>%
  align(align = "center", part = "header") %>%
  align(align = "center", part = "body", j = 2:7) %>%
  flextable::footnote(
    i = 1,
    j = 7,
    part = "header",
    value = as_paragraph("Serum soluble mediators were compared between clusters using the Kruskal-Wallis rank sum test."),
    ref_symbols = c("a")) %>%
  save_as_pptx(path = "tbls4.pptx")
```

```{r}
sessioninfo::session_info()
```
