---
title: "RNAseq analysis"
author: "Miles Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 1500px !important;
      width: 1500px !important;
    }
    body {
      max-width: 1500px !important;
    }
    caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
```

```{r setup, include=FALSE}
options(width = 1200)
knitr::opts_chunk$set(echo = FALSE)
theme_set(theme_cowplot())
loadd(annotation_info,
  group_pal,
  module_scores_with_viral,
  ifn_modules,
  inflame_modules,
  module_ISGs,
  pca_qc,
  pca_results,
  removed_outliers,
  res,
  sample_dendrogram,
  sample_dists,
  sampleDistMatrix,
  up_table,
  down_table,
  top_up,
  top_down,
  tx_files,
  umap_results,
  detected_viral_transcripts)
```


```{r}
loadd(md)
md %>%
  tabyl(disease_class) %>%
  adorn_pct_formatting(digits = 2) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, bold = TRUE, border_right = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = FALSE, width = "10em")

```

# Differential Gene Expression {.tabset .tabset-fade .tabset-pills}
```{r}
res %>% 
  map(alt_summary) %>%
  enframe() %>%
  unnest(cols = c(value)) %>%
  mutate(`up %` = round(100*(as.integer(up)/as.integer(total)),
                        digits = 2) %>% paste("%"),
         `down %` = round(100*(as.integer(down)/as.integer(total)),
                        digits = 2) %>% paste("%"),
         `low counts %` = round(100*(as.integer(lowcounts)/as.integer(total)),
                                digits = 2) %>% paste("%"),
         name = str_replace_all(string = name, 
                                pattern = "_",
                                replacement = " ")) %>%
  select(comparison = name, up, `up %`, down, `down %`, `mean count cutoff` = ft, `low counts` = lowcounts, `low counts %`) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  footnote(paste(alt_summary(res[[1]])$total, "with nonzero total read count", "\nOutliers: ", alt_summary(res[[1]])$outlier))
```

## 25 genes with largest negative log-fold change in experimental group
```{r echo = FALSE, results = "asis"}
loadd(down_tables)
for (i in seq_along(down_tables)){
  down_tables[[i]] %>%
    dplyr::select(-baseMean) %>%
    dplyr::filter(log2FoldChange > 0) %>%
    arrange(desc(log2FoldChange)) %>%
    mutate(
      gene = cell_spec(gene,
                       bold = TRUE),
      padj = cell_spec(padj,
                       color = "white",
                       bold = TRUE,
                       background = spec_color(padj)),
      log2FoldChange = color_bar("tomato")(log2FoldChange)) %>%
    ungroup() %>%
    rename(!!paste0("pvalue", footnote_marker_symbol(1)) := pvalue,
           #!!paste0("stat", footnote_marker_symbol(2)) := stat,
           !!paste0("padj", footnote_marker_symbol(2)) := padj,
           "lfc standard error" = lfcSE) %>%
    kable(escape = FALSE,
          caption = paste0("<b>", names(down_tables)[[i]],
          ": downregulated in ",
          str_split(string = names(down_tables)[[i]],
                    pattern = "_vs_",
                    simplify = TRUE)[[1]], "</b>")) %>%
    kable_styling(
      bootstrap_options = c("striped", "hover", "condensed", "responsive"),
      full_width = TRUE,
      fixed_thead = TRUE) %>%
    #footnote(symbol = c("Wald test p-values", "Wald statistic", "Benjamini–Hochberg adjusted value")) %>%
    footnote(symbol = c("Wald test p-values", "Benjamini–Hochberg adjusted value")) %>%
    column_spec(2, color = "black") %>%
    print()
}
```

## 25 genes with largest positive log-fold change in experimental group
```{r echo = FALSE, results = "asis"}
loadd(up_tables)
for (j in seq_along(up_tables)){
  up_tables[[j]] %>%
    select(-baseMean) %>%
    filter(log2FoldChange > 0) %>%
    arrange(desc(log2FoldChange)) %>%
    mutate(
      gene = cell_spec(gene,
                       bold = TRUE),
      padj = cell_spec(padj,
                       color = "white",
                       bold = ifelse(pvalue < 0.05, T, F),
                       background = spec_color(padj)),
      log2FoldChange = color_bar("lightgreen")(log2FoldChange)) %>%
    ungroup() %>%
    rename(!!paste0("pvalue", footnote_marker_symbol(1)) := pvalue,
           #!!paste0("stat", footnote_marker_symbol(2)) := stat,
           !!paste0("padj", footnote_marker_symbol(2)) := padj,
           "lfc standard error" = lfcSE) %>%
    kable(escape = FALSE,
          caption = paste0("<b>", names(up_tables)[[i]],
          ": upregulated in ",
          str_split(string = names(up_tables)[[i]],
                    pattern = "_vs_",
                    simplify = TRUE)[[1]], "</b>")) %>%
    kable_styling(
      bootstrap_options = c("striped", "hover", "condensed", "responsive"),
      full_width = TRUE,
      fixed_thead = TRUE) %>%
    #footnote(symbol = c("Wald test p-values", "Wald statistic", "Benjamini–Hochberg adjusted value")) %>%
    footnote(symbol = c("Wald test p-values", "Benjamini–Hochberg adjusted value")) %>%
    column_spec(2, color = "black") %>% print()
  }
```

## Expression of top class DE genes by class
```{r, fig.width=12, fig.height=9}
loadd(deg_class, final_md, vsd_exprs)

deg_class_tbl <- as_tibble(deg_class, rownames = "gene")
comparisons = unique(deg_class_tbl$comparison)

final_md_tbl <- final_md %>% as_tibble(rownames = "subject") %>% filter(subject %in% colnames(vsd_exprs))

range_bound = 4
zscore_range <- c(-range_bound:range_bound)
for (i in names(up_tables)){
  #comparison_genes <- filter(deg_class_tbl, comparison == i) %>% pull(gene)
  comparison_genes <- c(pull(up_tables[[i]], gene), pull(down_tables[[i]], gene))
  people <- filter(final_md_tbl, disease_class %in% str_split(i, "_vs_", simplify = TRUE)) %>% pull(subject)
    
  t(vsd_exprs)[people,comparison_genes] %>% 
      pheatmap(
        scale = 'column',
        main = i,
        fontsize = 9,
        fontsize_col = 9,
        show_rownames = FALSE,
        annotation_row = annotation_info,
        annotation_colors = group_pal,
        angle_col = 315,
        breaks = zscore_range,
        color = viridis(n = length(zscore_range), option = "D"),
        border_color = NA
      )
  }
```

## Expression of all top DE genes
```{r, fig.width=12, fig.height=9}
loadd(vsd_exprs, deg_means)
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
t(vsd_exprs)[,colnames(deg_means)] %>% 
    pheatmap(
      scale = 'column',
      fontsize = 9,
      fontsize_col = 6,
      show_rownames = FALSE,
      annotation_row = annotation_info,
      annotation_colors = group_pal,
      angle_col = 315,
      breaks = zscore_range,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      gaps_col = class_gaps,
      border_color = NA
      )
```

## Volcano plot of gene expression
```{r fig.height=6, fig.width=9}

  for(k in seq_along(res)){
      volcano_data <- 
        res[[k]] %>%
        as_tibble(rownames = "gene") %>%
        mutate(sig = case_when(
          padj > 0.01 ~ "insig",
          padj < 0.01 & log2FoldChange <= -0.25 ~ "down",
          padj < 0.01 & log2FoldChange >= 0.25 ~ "up",
          padj < 0.01 & between(log2FoldChange, -1, 1) ~ "ignore",
          is.na(padj) ~ "insig")
          )
      

  top_volcano_data =
    rbind(
      top_n(filter(volcano_data, sig == "up"), n = 20, log2FoldChange),
      top_n(filter(volcano_data, sig == "down"), n = -20, log2FoldChange)
    )
    
  print(ggplot(data = volcano_data,
         mapping = aes(x = log2FoldChange,
                       y = 1/padj,
                       color = sig)) +
    geom_point(alpha = 0.25) +
    scale_y_log10() +
    theme_cowplot() +
    scale_color_manual(values = c("down" = 'red',
                                  "up" = "blue",
                                  "insig" = "black",
                                  "ignore" = 'grey')) +
    geom_text_repel(data = top_volcano_data,
                    aes(label = gene)) +
    guides(color = "none") +
    labs(title = names(res)[[k]]))
  }
```

## Expression of selected interferon-stimulated genes
```{r, fig.width=9, fig.height=6}
loadd(ISGs)

range_bound = 4
zscore_range <- c(-range_bound:range_bound)
t(vsd_exprs)[,ISGs] %>%
    pheatmap(
      scale = 'column',
      fontsize = 6,
      show_rownames = FALSE,
      annotation_row = annotation_info,
      annotation_colors = group_pal,
      angle_col = 315,
      breaks = zscore_range,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      gaps_col = class_gaps,
      border_color = NA)

rm(ISGs)
```

## Expression of individual genes in the M1.2, M3.4, and M5.12 gene modules
```{r fig.height=6, fig.width=9}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
vsd_exprs %>%
  as_tibble(rownames = "gene") %>%
  filter(gene %in% rownames(module_ISGs)) %>%
  filter(gene != "MT1A") %>%
  column_to_rownames("gene") %>%
  t() %>%
  `[`(,rownames(module_ISGs)[(rownames(module_ISGs) != "MT1A")]) %>%
  pheatmap(
    scale = 'column',
    fontsize = 6,
    show_rownames = FALSE,
    annotation_row = annotation_info,
    annotation_col = module_ISGs,
    annotation_colors = group_pal,
    cluster_cols = FALSE,
    angle_col = 315,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA)
```

# Sample Clustering

Using gene expression distance, a sample adjacency matrix was calculated from a shared nearest neighbor graph.  This, in turn, was used to cluster the samples, with the optimal clustering resolution being determined by silhouette width at multiple resolutions.

```{r}
annotation_info %>%
  mutate(cluster = as_factor(cluster)) %>%
  group_by(disease_class,
           cluster) %>%
  summarise(n = n()) %>%
  pivot_wider(names_from = cluster,
              values_from = n) %>%
  DT::datatable()

annotation_info %>%
  mutate(cluster = as_factor(cluster)) %>%
  select(disease_class, cluster) %>%
  group_by(disease_class) %>%
  mutate(total_disease = n()) %>%
  ungroup() %>%
  group_by(disease_class, cluster) %>%
  mutate(total_cluster_disease = n(),
         freq = total_cluster_disease/total_disease) %>%
  distinct() %>%
  ggplot(aes(
    x = disease_class,
    y = freq,
    group = cluster,
    fill = cluster
  )) +
  geom_col() +
  labs(y = "Proportion of disease class in cluster") +
  scale_fill_paletteer_d("ggthemes::calc") +
  theme_cowplot()
```


# Clustering and variation of samples in reduced dimensional space {.tabset .tabset-fade .tabset-pills}

## PCA

### PCA, by disease classification
```{r, fig.width=12, fig.height=6}
p1 <-
  ggplot(
    data = pca_results,
    aes(
      x = PC1,
      y = PC2,
      fill = disease_class,
      line = 1
    ),
    shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    theme_cowplot() +
    theme(legend.position = "none")

p2 <-
  ggplot(
    data = pca_results,
    aes(
      x = PC1,
      y = PC2,
      fill = disease_class,
      line = 1
      ), shape = 1
    ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  scale_fill_brewer(palette = "Set1") +
  scale_color_brewer(palette = "Set1") +
  theme_cowplot() +
  geom_mark_ellipse(aes(fill = disease_class,
                        color = disease_class),
                    alpha = 0.05
                    ) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```


### PCA, by plate
```{r, fig.width=12, fig.height=7}
p1 <- ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = run_id
    )
  ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  labs(color = "run_id") +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv")

p2 <- ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = run_id
    )
  ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  labs(color = "run_id") +
  geom_mark_ellipse(aes(
    fill = run_id,
    color = run_id
    ),
    alpha = 0.05
    ) +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv") +
  guides(color = guide_legend(nrow = 4)) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```

### PCA, by initial sample concentration
```{r, fig.width=8, fig.height=6}
ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = initial_concentration_ng_ul
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  scale_fill_gradient2(low = "blue",
                       mid = "grey90",
                       high = "red",
                       midpoint = 10)

```

### PCA, by cluster
```{r, fig.width=7, fig.height=6}
# ggplot(
#   data = pca_results,
#   mapping = aes(
#     x = PC1,
#     y = PC2,
#     fill = cluster
#   )
# ) +
#   geom_point(
#     shape = 21,
#     colour = "black",
#     size = 2,
#     stroke = 0.5,
#     alpha = 0.6
#   ) +
#   scale_fill_manual(values = group_pal$cluster)
pca_results %>%
  plot_ly(x = ~PC1,
          y = ~PC2,
          color = ~cluster,
          alpha = 0.75,
          colors = as.character(paletteer_d("ggthemes::calc")),
          text = ~paste('<br>Sample: ', sample_name,
                        '<br>Classification: ', disease_class,
                        '<br>Ethnicity ', race_code,
                        '<br>Cluster: ', cluster,
                        '<br>Sex: ', sex,
                        '<br>Plate: ', run_id
          )
  ) %>%
  add_markers(marker = list(size = 7.5,
                            line = list(color = 'rgba(0, 0, 0, .8)',
                                        width = 1)))
```

### 3D PCA, colored by cluster
```{r}
pca_results %>%
  plot_ly(x = ~PC1,
          y = ~PC2,
          z = ~PC3,
          color = ~cluster,
          size = 2.5,
          alpha = 0.9,
          colors = as.character(paletteer_d("ggthemes::calc")),
          text = ~paste('<br>Sample: ', sample_name,
                        '<br>Classification: ', disease_class,
                        '<br>Ethnicity ', race_code,
                        '<br>Cluster: ', cluster,
                        '<br>Sex: ', sex,
                        '<br>Plate: ', run_id
          )
  ) %>%
  add_markers(marker = list(size = 3,
                            line = list(color = 'rgba(0, 0, 0, .8)',
                                        width = 1.5)))
```


## UMAP

### UMAP, by disease classification
```{r, fig.width=12, fig.height=7}

p1 <-
  ggplot(
    data = umap_results,
    aes(
      x = umap_1,
      y = umap_2,
      fill = disease_class,
      line = 1
    ),
    shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    theme_cowplot() +
    theme(legend.position = "none")

p2 <-
  ggplot(
    data = umap_results,
    aes(
      x = umap_1,
      y = umap_2,
      fill = disease_class,
      line = 1
    ), shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    scale_color_brewer(palette = "Set1") +
    theme_cowplot() +
    geom_mark_ellipse(aes(fill = disease_class,
                          color = disease_class),
                      alpha = 0.05
    ) +
  guides(color = guide_legend(nrow = 4)) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```

### UMAP, by plate
```{r, fig.width=12, fig.height=7}
p1 <- ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = run_id
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  labs(color = "run_id") +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv")


p2 <- ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = run_id
    )
  ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  labs(color = "run_id") +
  geom_mark_ellipse(aes(
    fill = run_id,
    color = run_id
    ),
    alpha = 0.05
    ) +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv") +
  guides(color = guide_legend(nrow = 4)) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```

### UMAP, by initial sample concentration
```{r, fig.width=8, fig.height=6}
ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = initial_concentration_ng_ul
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  scale_fill_gradient2(low = "blue",
                       mid = "grey90",
                       high = "red",
                       midpoint = 10)
```

### UMAP, by cluster
```{r, fig.width=7, fig.height=6}
# ggplot(
#   data = umap_results,
#   mapping = aes(
#     x = umap_1,
#     y = umap_2,
#     fill = cluster
#   )
# ) +
#   geom_point(
#     shape = 21,
#     colour = "black",
#     size = 2,
#     stroke = 0.5,
#     alpha = 0.6
#   ) +
#   scale_fill_manual(values = group_pal$cluster)

umap_results %>%
  plot_ly(x = ~umap_1,
          y = ~umap_2,
          color = ~cluster,
          alpha = 0.75,
          colors = group_pal$cluster,
          text = ~paste('<br>Sample: ', sample_name,
                        '<br>Classification: ', disease_class,
                        '<br>Ethnicity ', race_code,
                        '<br>Cluster: ', cluster,
                        '<br>Sex: ', sex,
                        '<br>Plate: ', run_id
          )
  ) %>%
  add_markers(marker = list(size = 7.5,
                            line = list(color = 'rgba(0, 0, 0, .8)',
                                        width = 1)))
```

### UMAP, in 3D, colored by cluster
```{r}
umap_results %>%
  plot_ly(x = ~umap_1,
          y = ~umap_2,
          z = ~umap_3,
          color = ~cluster,
          size = 2.5,
          alpha = 0.9,
          colors = as.character(paletteer_d("ggthemes::calc")),
          text = ~paste('<br>Sample: ', sample_name,
                        '<br>Classification: ', disease_class,
                        '<br>Ethnicity ', race_code,
                        '<br>Cluster: ', cluster,
                        '<br>Sex: ', sex,
                        '<br>Plate: ', run_id
          )
  ) %>%
  add_markers(marker = list(size = 3,
                            line = list(color = 'rgba(0, 0, 0, .8)',
                                        width = 1.5)))
```

# Module scores {.tabset .tabset-fade .tabset-pills}

## SLE Interferon module scores
```{r fig.height=9, fig.width=4}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name, one_of(ifn_modules)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## SLE Inflammatory module scores
```{r fig.height=9, fig.width=6}
range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name, one_of(inflame_modules)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## Low-density granulocyte module scores
```{r fig.height=9, fig.width=5}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(names(ldg_modules))) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## SLE Modules with an annotated associated pathway
```{r fig.height=9, fig.width=11}
loadd(annotated_modules)

range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(annotated_modules$module)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale = 'column',
    fontsize = 9,
    show_rownames = FALSE,
    annotation_row = annotation_info,
    annotation_col = column_to_rownames(annotated_modules, var = "module"),
    annotation_colors = group_pal,
    angle_col = 315,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA)

```

## All SLE modules
```{r fig.height=9, fig.width=11}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         matches("^M[[:digit:]]+")) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(scale = 'column',
           fontsize = 9,
           annotation_row = annotation_info,
           annotation_col = module_annot,
           annotation_colors = group_pal %>%
             `$<-`(type,
                   c(.$type,
                     Undetermined = "#000000")),
           show_rownames = FALSE,
           show_colnames = FALSE,
           angle_col = 315,
           breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA)
```

# Viral trancription {.tabset .tabset-fade .tabset-pills}

Samples were examined for transcripts from the following viruses:
HSV-1, HSV-2, VZV, EBV, EBV-2, HCMV, HHV7, KSHV, Adenovirus 2, B19, HIV-1, HBV, HPV-2, HPV-16, JCV, and MCV-1

## Detected viral transcription and Interferon module scores
```{r fig.height=10, fig.width=10}
range_bound = 5
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(ifn_modules,
                detected_viral_transcripts)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA,
    angle=315)
```

## Detected viral transcription and Inflammatory module scores
```{r fig.height=10, fig.width=9}
range_bound = 5
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(inflame_modules,
                detected_viral_transcripts)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(scale = "column",
           annotation_colors = group_pal,
           annotation_row = annotation_info,
           breaks = zscore_range,
           fontsize = 9,
           show_rownames = FALSE,
           color = viridis(n = length(zscore_range)-1, option = "D"),
           border_color = NA,
           angle=315)
```

# WGCNA

## Expression similarity dendrogram
```{r echo=FALSE, fig.height=6, fig.width=10}
loadd(wgcna_modules)
plotDendroAndColors(dendro = wgcna_modules$dendrograms[[1]],
                    colors = wgcna_modules$colors[wgcna_modules$blockGenes[[1]]],
                    groupLabels = "Module colors",
                    main = "Gene dendrogram and module colors in block 1",
                    dendroLabels = FALSE,
                    hang = 0.03,
                    addGuide = TRUE,
                    guideHang = 0.05)
```

## Expression of identified modules {.tabset .tabset-fade .tabset-pills}

### By cluster
```{r echo=FALSE, fig.height=8, fig.width=12}
loadd(wgcna_scores)
loadd(cluster_pal)

module_scores_with_viral %>%
  select(cluster,
         starts_with("ME")) %>%
  pivot_longer(-cluster,
               names_to = "module",
               values_to = "score") %>%
  mutate(cluster = as_factor(cluster)) %>%
  ggplot(aes(x = cluster,
             y = score,
             fill = cluster)) +
  geom_violin() +
  geom_jitter(size = 0.5, alpha = 0.25) +
  scale_fill_manual(values = cluster_pal) +
  facet_wrap(facets = vars(module),
             scales = "free_y") +
  theme_cowplot()
```

### By disease classification
```{r echo=FALSE, fig.height=8, fig.width=12}
module_scores_with_viral %>%
  select(disease_class,
         starts_with("ME")) %>%
  pivot_longer(-disease_class,
               names_to = "module",
               values_to = "score") %>%
  mutate(disease_class = as_factor(disease_class)) %>%
  ggplot(aes(x = disease_class,
             y = score,
             fill = disease_class)) +
  geom_violin() +
  geom_jitter(size = 0.5, alpha = 0.25) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  facet_wrap(facets = vars(module),
             scales = "free_y") +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1))
```

## WGCNA eigenvalues
```{r fig.height=9, fig.width=11}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         matches("^ME")) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(scale = 'column',
           fontsize = 9,
           annotation_row = annotation_info,
           annotation_colors = group_pal %>%
             `$<-`(type,
                   c(.$type,
                     Undetermined = "#000000")),
           show_rownames = FALSE,
           show_colnames = TRUE,
           angle_col = 315,
           breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA)
```

## Gene ontology GSEA of WGCNA modules
```{r fig.height=11, fig.width=12}
loadd(MEplotting)

for(i in 1:ceiling(length(unique(MEplotting$module))/6)){
    
      plots <- ggplot(data = MEplotting,
               mapping = aes(
                 y = GeneRatio,
                 x = order,
                 color = p.adjust)
               ) +
          geom_point(stat = "identity",
                     size = 2) +
          scale_x_continuous(breaks = MEplotting$order,
                           labels = str_wrap(MEplotting$ID, width = 40),
                           expand = c(0.1,0.1)) +
          scale_color_gradient(low = "blue", high = "red") +
          labs(color = "adjusted\np value") +
          facet_wrap_paginate(~ module,
                              ncol = 2,
                              nrow = 3,
                              scales = "free",
                              page = i) +
          theme(axis.text.y = element_text(size = 9),
                axis.text.x = element_text(size = 9),
                panel.background = element_rect(fill = "white", linetype = "solid"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                panel.grid = element_line(color = "grey90"),
                axis.title.y = element_blank()) +
          coord_flip()
      print(plots)
}
```


## Use of WGCNA modules in classifying clusters.

The module eigengene scores were used to train a random forest model with repeated cross validation to predict cluster identity.

The model was trained with 75% of the dataset and tested on 25%.

```{r}
loadd(wgcna_cluster_rf_cv, wgcna_cluster_test)

caret::confusionMatrix(
  data = predict(wgcna_cluster_rf_cv, wgcna_cluster_test),
  reference = as_factor(wgcna_cluster_test$cluster)
)
```

The relative importance of each eigengene in classification:

```{r}
loadd(wgcna_cluster_rf_cv_varImp)

wgcna_cluster_rf_cv_varImp
```

## Use of WGCNA modules in labeling disease class.

The module eigengene scores were used to train a random forest model with repeated cross validation to predict cluster identity.

The model was trained with 75% of the dataset and tested on 25%.

```{r}
loadd(wgcna_disease_class_rf_cv, wgcna_disease_class_test)

caret::confusionMatrix(
  data = predict(wgcna_disease_class_rf_cv, wgcna_disease_class_test),
  reference = as_factor(wgcna_disease_class_test$disease_class)
)
```

The relative importance of each eigengene in classification:

```{r}
loadd(wgcna_disease_class_rf_cv_varImp)

wgcna_disease_class_rf_cv_varImp
```

# Data processing of RNAseq data

## Overview of sample characteristics:

```{r fig.height=4, fig.width=10}
loadd(md_cat_data)
inspectdf::show_plot(md_cat_data, high_cardinality=1, col_palette = 3) +
  labs(title = "Frequency of categorical levels")
```

```{r fig.height=4, fig.width=10}
loadd(md_num_data)
inspectdf::show_plot(md_num_data, col_palette = 3) + labs(title = "Distribution of sample characteristic values")
rm(md_num_data)
```

```{r echo=FALSE}
message(str_glue("{length(md$sample_name)} samples selected for analysis"))
message(str_glue("{length(names(tx_files))} files found"))
if (!all(names(tx_files) %in% md$sample_name)){
  message("The following files do not have a matching metadata entry:")
  print(names(tx_files)[(names(tx_files) %nin% md$sample_name)])
} else {
  message("All files match a metadata entry.")
}

if (!all(md$sample_name %in% names(tx_files))){
  message("A count file was not found for the following selected samples:")
  print(md$sample_name[(md$sample_name %nin% names(tx_files))])
} else {
  message("All expected files found.")
}
```

## Sample QC filtering

Samples with a PC1 > 2 (the red lines) and/or PC2 zscore > 2.5 (the green lines) were dropped from analysis.
```{r, fig.width=6, fig.height=6}
vline <- function(x = 0, color = "red") {
  list(
    type = "line", 
    y0 = 0, 
    y1 = 1, 
    yref = "paper",
    x0 = x, 
    x1 = x, 
    line = list(color = color)
  )
}

hline <- function(y = 0, color = "green") {
  list(
    type = "line", 
    x0 = 0, 
    x1 = 1, 
    xref = "paper",
    y0 = y, 
    y1 = y, 
    line = list(color = color)
  )
}

pca_qc %>%
  plot_ly(
    type = "scattergl",
    mode = "markers",
    x = ~PC1,
    y = ~PC2,
    marker = list(size = 7.5,
                  color = "#AAAAAA",
                  line = list(color = 'rgba(0, 0, 0, .8)',
                              width = 1)),
    text = ~paste('<br>Subject: ', sample,
                  '<br>Sex: ', sex,
                  '<br>Ethnicity: ', race_code,
                  '<br>Disease class: ', disease_class,
                  '<br>Run: ', run_id),
    hovertemplate = '%{text}',
    showlegend = FALSE,
    alpha = 0.5,
    name = ""
  ) %>%
  layout(
    shapes = list(
      vline(pc1_zscore_threshold*sd(pca_qc$PC1) + mean(pca_qc$PC1)),
      vline(-pc1_zscore_threshold*sd(pca_qc$PC1) + mean(pca_qc$PC1)),
      
      hline(pc2_zscore_threshold*sd(pca_qc$PC2) + mean(pca_qc$PC2)),
      hline(-pc2_zscore_threshold*sd(pca_qc$PC2) + mean(pca_qc$PC2))
      )
    )
# pca_qc %>% ggplot(aes(x = PC1,
#                       y = PC2)) +
#   geom_point(shape = 21,
#              colour = "black",
#              fill = "grey80",
#              size = 2,
#              stroke = 0.5,
#              alpha = 0.6) +
#   geom_vline(aes(xintercept = (pc1_zscore_threshold*sd(.data$PC1)) + mean(.data$PC1)),
#              color = "red") +
#   geom_vline(aes(xintercept = -(pc1_zscore_threshold*sd(.data$PC1)) + mean(.data$PC1)),
#              color = "red") +
#   geom_hline(aes(yintercept = (pc2_zscore_threshold*sd(.data$PC2)) + mean(.data$PC2)),
#              color = "green") +
#   geom_hline(aes(yintercept = -(pc2_zscore_threshold*sd(.data$PC2)) + mean(.data$PC2)),
#              color = "green")
```

```{r}
if (length(removed_outliers) > 0) {
  filter(md, sample_name %in% removed_outliers) %>%
  kable(caption = "Removing the following samples:") %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = FALSE)
}
```

### Overall correlation between samples
```{r, fig.width=12, fig.height=9}
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sample_dists,
         clustering_distance_cols = sample_dists, 
         annotation_row = annotation_info,
         show_rownames = FALSE,
         show_colnames = TRUE,
         annotation_colors = group_pal,
         angle_col = 45)
```

# Data quality

## MA-plot

This MA plot is of shrunken log2 fold changes (using the *apeglm* package), in which actual fold changes are adjusted to compensate for the high variablity of genes with few counts.

Dots in red indicate a significant expression difference between Control and SLE.
```{r echo=FALSE}
loadd(res)
#l <- tagList()
for(m in seq_along(res)){
  plotMA(res[[m]], ylim = c(-6,6), main = names(res)[[m]], colSig = "red")
  # l[[m]] <- res[[m]] %>%
  # as_tibble(rownames = "gene") %>%
  # mutate(
  #   sig = case_when(
  #     padj < 0.05 ~ TRUE,
  #     padj >= 0.05 ~ FALSE,
  #     is.na(padj) ~ FALSE
  #     ),
  # ) %>%
  # plot_ly(
  #   type = "scattergl",
  #   x = ~baseMean,
  #   y = ~log2FoldChange,
  #   color = ~sig,
  #   alpha = 0.75,
  #   colors = c("grey30","red"),
  #   text = ~paste('<br>Gene: ', gene,
  #                 '<br>LFC: ', signif(log2FoldChange, digits = 2),
  #                 '<br>p adjusted: ', scientific(padj, digits = 2))
  #   ) %>%
  #   add_markers(marker = list(size = 7.5,
  #                             line = list(color = 'rgba(0, 0, 0, .8)',
  #                                         width = 1))) %>%
  #   layout(xaxis = list(type = "log"))
}
#l
#plotMA(res[[1]], ylim = c(-4,4))
```

## Dispersion estimates
```{r}
loadd(disp_plot)
disp_plot
rm(disp_plot)
```

<!-- ## Cluster resolution optimization -->

<!-- The samples are clustered using the Leiden algorithm on shared nearest neighbors calculated from sample distances.  Since the clustering resolution is an emperical value that has to be determined for each dataset, a range of resolutions were tested and then the silhouette coefficient was calculated at each resolution.  A higher coefficient means better intracluster simliarity and intercluster differences. -->
<!-- ```{r} -->
<!-- loadd(cluster_res_scan) -->
<!-- plot_resolution_silhouette_coeff(cluster_res_scan) -->
<!-- ``` -->

## K-means clustering optimization

To determine the optimal *k*, a silhouette coefficient was calculated for *k* = (3,20) (*k* = 2 was ignored because it just returns the two disease clusters).  A higher coefficient means better intracluster simliarity and intercluster differences.

```{r}
loadd(sample_cluster_info)

sample_cluster_info[["avg_sils"]] %>%
  filter(name > 2) %>%
  ggplot(
    aes(
      x = as.integer(name),
      y = value
    )
  ) +
  geom_line() +
  geom_label_repel(
    data = top_n(
      x = filter(
        sample_cluster_info[["avg_sils"]],
        name > 2),
      n = 1,
      wt = value),
    mapping = aes(
      x = name,
      y = value,
      label = name
      ), 
    arrow = arrow(length = unit(0.02, "npc")),
    box.padding = 1,
    point.padding = 0.5,
    nudge_y = 1,
    nudge_x = 0.5
    ) +
  scale_x_continuous(breaks = sample_cluster_info[["avg_sils"]]$name) +
  labs(x = ~italic(k)~" number of clusters",
       y = "Silhouette coefficient") +
  theme_cowplot()
```


# Sample Sanity check

To determine if there is a gross mixup of samples, examine the expression of sex-specific transcripts to ensure that only female-subject samples express "XIST" and male-subject samples "DDX3Y:

```{r}
sex_specific_genes <- tibble(chr = c(rep("Y",16), rep("X",2)),
                             symbol = c("DDX3Y","EIF1AY",
                                       "KDM5D","NLGN4Y",
                                       "PCDH11Y","PRORY",
                                       "PRY","PRY2",
                                       "RPS4Y1","RPS4Y2",
                                       "TBL1Y","TMSB4Y",
                                       "USP9Y","UTY",
                                       "VCY","ZFY",
                                       "XIST","ZFX")) %>%
  filter(symbol %in% rownames(vsd_exprs))

range_bound = 5
zscore_range <- c(-range_bound:range_bound)

t(vsd_exprs)[,c(sex_specific_genes[['symbol']])] %>%
  pheatmap(
    scale='column',
    fontsize = 6,
    show_rownames = FALSE,
    annotation_row = annotation_info[,c("sex", "disease_class")],
    annotation_colors = group_pal,
    annotation_col = column_to_rownames(sex_specific_genes, "symbol"),
    cluster_cols = FALSE,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA,
    angle=315)
```

These appear to be correct.

```{r}
sessioninfo::session_info()
```
