---
title: "ABC-MESS RNAseq analysis"
author: "Miles Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 1500px !important;
      width: 1500px !important;
    }
    body {
      max-width: 1500px !important;
    }
    caption {
      color: black;
      font-weight: bold;
      font-size: 1.0em;
    }
```

```{r setup, include=FALSE}
options(width = 1200)
knitr::opts_chunk$set(echo = FALSE)
theme_set(theme_cowplot())
```


```{r disease_class table}
loadd(dds_with_scores, md)
md <- colData(dds_with_scores) %>% as_tibble(rownames = "sample_name") %>% select(one_of(colnames(md)))
md %>%
  tabyl(disease_class) %>%
  adorn_pct_formatting(digits = 2) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE) %>%
  column_spec(1, bold = TRUE, border_right = TRUE, width = "10em") %>%
  column_spec(2, border_right = TRUE, width = "10em") %>%
  column_spec(3, border_right = FALSE, width = "10em")
```

# Differential Gene Expression {.tabset .tabset-fade .tabset-pills}
```{r DE table}
loadd(res)
res %>% 
  map(alt_summary) %>%
  enframe() %>%
  unnest(cols = c(value)) %>%
  mutate(`up %` = round(100*(as.integer(up)/as.integer(total)),
                        digits = 2) %>% paste("%"),
         `down %` = round(100*(as.integer(down)/as.integer(total)),
                        digits = 2) %>% paste("%"),
         `low counts %` = round(100*(as.integer(lowcounts)/as.integer(total)),
                                digits = 2) %>% paste("%"),
         name = str_replace_all(string = name, 
                                pattern = "_",
                                replacement = " ")) %>%
  select(comparison = name, up, `up %`, down, `down %`, `mean count cutoff` = ft, `low counts` = lowcounts, `low counts %`) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  footnote(paste(alt_summary(res[[1]])$total, "with nonzero total read count", "\nOutliers: ", alt_summary(res[[1]])$outlier))
```

## 25 genes with largest negative log-fold change in experimental group
```{r 25 top up DE, echo = FALSE, results = "asis"}
loadd(down_tables)
for (i in seq_along(down_tables)){
  down_tables[[i]] %>%
    dplyr::select(-baseMean) %>%
    dplyr::filter(log2FoldChange > 0) %>%
    arrange(desc(log2FoldChange)) %>%
    mutate(
      gene = cell_spec(gene,
                       bold = TRUE),
      padj = cell_spec(padj,
                       color = "white",
                       bold = TRUE,
                       background = spec_color(padj)),
      log2FoldChange = color_bar("tomato")(log2FoldChange)) %>%
    ungroup() %>%
    rename(!!paste0("pvalue", footnote_marker_symbol(1)) := pvalue,
           #!!paste0("stat", footnote_marker_symbol(2)) := stat,
           !!paste0("padj", footnote_marker_symbol(2)) := padj,
           "lfc standard error" = lfcSE) %>%
    kable(escape = FALSE,
          caption = paste0("<b>", names(down_tables)[[i]],
          ": downregulated in ",
          str_split(string = names(down_tables)[[i]],
                    pattern = "_vs_",
                    simplify = TRUE)[[1]], "</b>")) %>%
    kable_styling(
      bootstrap_options = c("striped", "hover", "condensed", "responsive"),
      full_width = TRUE,
      fixed_thead = TRUE) %>%
    #footnote(symbol = c("Wald test p-values", "Wald statistic", "Benjamini–Hochberg adjusted value")) %>%
    footnote(symbol = c("Wald test p-values", "Benjamini–Hochberg adjusted value")) %>%
    column_spec(2, color = "black") %>%
    print()
}
```

## 25 genes with largest positive log-fold change in experimental group
```{r 25 top down DE, echo = FALSE, results = "asis"}
loadd(up_tables)
for (j in seq_along(up_tables)){
  up_tables[[j]] %>%
    select(-baseMean) %>%
    filter(log2FoldChange > 0) %>%
    arrange(desc(log2FoldChange)) %>%
    mutate(
      gene = cell_spec(gene,
                       bold = TRUE),
      padj = cell_spec(padj,
                       color = "white",
                       bold = ifelse(pvalue < 0.05, T, F),
                       background = spec_color(padj)),
      log2FoldChange = color_bar("lightgreen")(log2FoldChange)) %>%
    ungroup() %>%
    rename(!!paste0("pvalue", footnote_marker_symbol(1)) := pvalue,
           #!!paste0("stat", footnote_marker_symbol(2)) := stat,
           !!paste0("padj", footnote_marker_symbol(2)) := padj,
           "lfc standard error" = lfcSE) %>%
    kable(escape = FALSE,
          caption = paste0("<b>", names(up_tables)[[j]],
          ": upregulated in ",
          str_split(string = names(up_tables)[[j]],
                    pattern = "_vs_",
                    simplify = TRUE)[[1]], "</b>")) %>%
    kable_styling(
      bootstrap_options = c("striped", "hover", "condensed", "responsive"),
      full_width = TRUE,
      fixed_thead = TRUE) %>%
    #footnote(symbol = c("Wald test p-values", "Wald statistic", "Benjamini–Hochberg adjusted value")) %>%
    footnote(symbol = c("Wald test p-values", "Benjamini–Hochberg adjusted value")) %>%
    column_spec(2, color = "black") %>% print()
  }
```

## Expression of top class DE genes by class
```{r top class DE, fig.width=12, fig.height=9}
loadd(deg_class, final_md, vsd_exprs, annotation_info, group_pal)

deg_class_tbl <- as_tibble(deg_class, rownames = "gene")
comparisons = unique(deg_class_tbl$comparison)

final_md_tbl <- final_md %>% as_tibble(rownames = "subject") %>% filter(subject %in% colnames(vsd_exprs))

range_bound = 4
zscore_range <- c(-range_bound:range_bound)
for (i in names(up_tables)){
  #comparison_genes <- filter(deg_class_tbl, comparison == i) %>% pull(gene)
  comparison_genes <- c(pull(up_tables[[i]], gene), pull(down_tables[[i]], gene))
  people <- filter(final_md_tbl, disease_class %in% str_split(i, "_vs_", simplify = TRUE)) %>% pull(subject)
    
  t(vsd_exprs)[people,comparison_genes] %>% 
      pheatmap(
        scale = 'column',
        main = i,
        fontsize = 9,
        fontsize_col = 9,
        show_rownames = FALSE,
        annotation_row = annotation_info,
        annotation_colors = group_pal,
        angle_col = 315,
        breaks = zscore_range,
        color = viridis(n = length(zscore_range), option = "D"),
        border_color = NA
      )
  }
```

## Expression of all top DE genes
```{r Expression of all top DE genes, fig.width=12, fig.height=9}
loadd(deg_means)
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
t(vsd_exprs)[,colnames(deg_means)] %>% 
    pheatmap(
      scale = 'column',
      fontsize = 9,
      fontsize_col = 6,
      show_rownames = FALSE,
      annotation_row = annotation_info,
      annotation_colors = group_pal,
      angle_col = 315,
      breaks = zscore_range,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      gaps_col = class_gaps,
      border_color = NA
      )
```

## Volcano plot of gene expression
```{r Volcano plot of gene expression, fig.height=6, fig.width=9}

l <- htmltools::tagList()

for(k in seq_along(res)){
 
    volcano_data <- 
      res[[k]] %>%
      as_tibble(rownames = "gene") %>%
      mutate(sig = case_when(
        padj > 0.01 ~ "insig",
        padj < 0.01 & log2FoldChange <= -0.25 ~ "down",
        padj < 0.01 & log2FoldChange >= 0.25 ~ "up",
        padj < 0.01 & between(log2FoldChange, -1, 1) ~ "ignore",
        is.na(padj) ~ "insig")
      )
    
    volcano_data_up =
      top_n(filter(volcano_data, sig == "up"), n = 20, log2FoldChange)
    volcano_data_down = 
        top_n(filter(volcano_data, sig == "down"), n = -20, log2FoldChange)
    
  l[[k]] <-
    volcano_data %>%
      plot_ly(
        x = ~log2FoldChange,
        y = ~(1/padj),
        color = ~sig,
        colors = c(
          "down" = 'red',
          "up" = "blue",
          "insig" = "black",
          "ignore" = 'grey'
          ),
        alpha = 0.25
      ) %>%
      add_markers(
        type = "scattergl",
        mode = "markers",
        text = c(volcano_data$gene),
        hovertemplate = 'Gene: <i>%{text}</i><br>log<sub>2</sub>Fold Change: %{x:.2f}<br>Adjusted p value: %{y:.2e}'
        ) %>%
      add_text(
        data = volcano_data_up,
        x = ~log2FoldChange,
        y = ~1/padj,
        text = ~gene,
        textposition = "top right",
        alpha = 1
        ) %>%
      add_text(
        data = volcano_data_down,
        x = ~log2FoldChange,
        y = ~1/padj,
        text = ~gene,
        textposition = "top left",
        alpha = 1
      ) %>%
      layout(
        title = str_replace_all(string = names(res)[[k]], pattern = "_", replacement = " "),
        yaxis = list(type = "log", title = "1/adjusted <i>p</i> value", exponentformat = "E"),
        xaxis = list(title = "log<sub>2</sub> fold change"),
        showlegend = FALSE
      )
}
l
```

## Expression of selected interferon-stimulated genes
```{r Expression of selected interferon-stimulated genes, fig.height=6, fig.width=9}
loadd(ISGs)

range_bound = 4
zscore_range <- c(-range_bound:range_bound)
t(vsd_exprs)[,ISGs] %>%
    pheatmap(
      scale = 'column',
      fontsize = 6,
      show_rownames = FALSE,
      annotation_row = annotation_info,
      annotation_colors = group_pal,
      angle_col = 315,
      breaks = zscore_range,
      color = viridis(n = length(zscore_range)-1, option = "D"),
      gaps_col = class_gaps,
      border_color = NA)

rm(ISGs)
```

## Expression of individual genes in the M1.2, M3.4, and M5.12 gene modules
```{r Expression of individual genes in the M1.2, M3.4, and M5.12 gene modules, fig.height=6, fig.width=9}
loadd(module_ISGs)
range_bound = 4
zscore_range <- c(-range_bound:range_bound)
vsd_exprs %>%
  as_tibble(rownames = "gene") %>%
  filter(gene %in% rownames(module_ISGs)) %>%
  filter(gene != "MT1A") %>%
  column_to_rownames("gene") %>%
  t() %>%
  `[`(,rownames(module_ISGs)[(rownames(module_ISGs) != "MT1A")]) %>%
  pheatmap(
    scale = 'column',
    fontsize = 6,
    show_rownames = FALSE,
    annotation_row = annotation_info,
    annotation_col = module_ISGs,
    annotation_colors = group_pal,
    cluster_cols = FALSE,
    angle_col = 315,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA)
```

# Sample Clustering

The eigenvales for the annotated Banchereau modules were used to calculate an adjacentcy matrix, which was in turn used to calculate the gap statistic to determine the optimal *k*-clusters.

```{r Sample Clustering chart}
annotation_info %>%
  group_by(disease_class,
           cluster) %>%
  summarise(n = n()) %>%
  pivot_wider(names_from = cluster,
              values_from = n) %>%
  DT::datatable()

annotation_info %>%
  select(disease_class, cluster) %>%
  group_by(disease_class) %>%
  mutate(total_disease = n()) %>%
  ungroup() %>%
  group_by(disease_class, cluster) %>%
  mutate(total_cluster_disease = n(),
         freq = total_cluster_disease/total_disease,
         cluster = as_factor(cluster)) %>%
  distinct() %>%
  ggplot(aes(
    x = disease_class,
    y = freq,
    group = cluster,
    fill = cluster
  )) +
  geom_col() +
  labs(y = "Proportion of disease class in cluster") +
  scale_fill_brewer(palette = "Set1") +
  theme_cowplot()
```

# Clustering and variation of samples in reduced dimensional space {.tabset .tabset-fade .tabset-pills}

## PCA

### PCA by disease classification
```{r PCA by disease classification, fig.width=12, fig.height=6}
loadd(pca_results)
p1 <-
  ggplot(
    data = pca_results,
    aes(
      x = PC1,
      y = PC2,
      fill = disease_class,
      line = 1
    ),
    shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    theme_cowplot() +
    theme(legend.position = "none")

p2 <-
  ggplot(
    data = pca_results,
    aes(
      x = PC1,
      y = PC2,
      fill = disease_class,
      line = 1
      ), shape = 1
    ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  scale_fill_brewer(palette = "Set1") +
  scale_color_brewer(palette = "Set1") +
  theme_cowplot() +
  geom_mark_ellipse(aes(fill = disease_class,
                        color = disease_class),
                    alpha = 0.05
                    ) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```


### PCA by plate
```{r PCA by plate, fig.width=12, fig.height=7}
p1 <- ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = run_id
    )
  ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  labs(color = "run_id") +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv")

p2 <- ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = run_id
    )
  ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  labs(color = "run_id") +
  geom_mark_ellipse(aes(
    fill = run_id,
    color = run_id
    ),
    alpha = 0.05
    ) +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv") +
  guides(color = guide_legend(nrow = 4)) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```

### PCA by ethnicity
```{r PCA by ethnicity, fig.width=12, fig.height=6}
p1 <-
  pca_results %>%
  select(PC1, PC2, sample_name, ethnicity) %>%
  mutate(ethnicity = as_factor(ethnicity)) %>%
  ggplot(
    aes(
      x = PC1,
      y = PC2,
      fill = ethnicity,
      line = 1
    ),
    shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_paletteer_d(palette = "ggthemes::calc") +
    theme_cowplot() +
    theme(legend.position = "none")

p2 <-
  pca_results %>%
  select(PC1, PC2, sample_name, ethnicity) %>%
  mutate(ethnicity = as_factor(ethnicity)) %>%
  ggplot(
    aes(
      x = PC1,
      y = PC2,
      fill = ethnicity,
      line = 1
      ), shape = 1
    ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  scale_fill_paletteer_d(palette = "ggthemes::calc") +
  scale_color_paletteer_d(palette = "ggthemes::calc") +
  theme_cowplot() +
  geom_mark_ellipse(aes(fill = ethnicity,
                        color = ethnicity),
                    alpha = 0.05
                    ) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```

### PCA by initial sample concentration
```{r, fig.width=8, fig.height=6}
ggplot(
  data = pca_results,
  mapping = aes(
    x = PC1,
    y = PC2,
    fill = initial_concentration_ng_ul
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  scale_fill_gradient2(
    low = "blue",
    mid = "grey90",
    high = "red",
    midpoint = 10
    ) +
  theme_cowplot()
```

### PCA by K-means cluster
```{r PCA by K-means cluster, fig.width=7, fig.height=6}
pca_results %>%
  plot_ly(
    x = ~PC1,
    y = ~PC2,
    text = str_glue("Sample: {.$sample_name}<br>Classification: {.$disease_class}<br>Ethnicity {.$ethnicity}<br>Cluster: {.$cluster}<br>Sex: {.$sex}")
  ) %>%
  add_markers(
    color = ~cluster,
    type = "scattergl",
    alpha = 0.75,
    mode = "markers",
    colors = group_pal$cluster,
    marker = list(
      size = 7.5,
      line = list(
        color = 'rgba(0, 0, 0, .8)',
        width = 1
      )
    ),
    hovertemplate = "%{text}"
  )
```

### 3D PCA colored by K-means cluster
```{r 3D PCA colored by K-means cluster}
pca_results %>%
  plot_ly(
    x = ~PC1,
    y = ~PC2,
    z = ~PC3,
    colors = group_pal$cluster
  ) %>%
  add_markers(
    text = str_glue("Sample: {pca_results$sample_name}<br>Classification: {pca_results$disease_class}<br>Ethnicity {pca_results$ethnicity}<br>Cluster: {pca_results$cluster}<br>Sex: {pca_results$sex}<br>Plate: {pca_results$run_id}"),
    color = ~cluster,
    size = 2.5,
    alpha = 0.9,
    hovertemplate = "%{text}",
    marker = list(
      size = 3,
      line = list(
        color = 'rgba(0, 0, 0, .8)',
        width = 1.5
      )
    )
  )
```


## UMAP

### UMAP by disease classification
```{r UMAP by disease classification, fig.width=12, fig.height=7}
loadd(umap_results)
p1 <-
  ggplot(
    data = umap_results,
    aes(
      x = umap_1,
      y = umap_2,
      fill = disease_class,
      line = 1
    ),
    shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    theme_cowplot() +
    theme(legend.position = "none")

p2 <-
  ggplot(
    data = umap_results,
    aes(
      x = umap_1,
      y = umap_2,
      fill = disease_class,
      line = 1
    ), shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_brewer(palette = "Set1") +
    scale_color_brewer(palette = "Set1") +
    theme_cowplot() +
    geom_mark_ellipse(aes(fill = disease_class,
                          color = disease_class),
                      alpha = 0.05
    ) +
  guides(color = guide_legend(nrow = 4)) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```

### UMAP by plate
```{r UMAP by plate, fig.width=12, fig.height=7}
p1 <- ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = run_id
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  labs(color = "run_id") +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv")


p2 <- ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = run_id
    )
  ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  labs(color = "run_id") +
  geom_mark_ellipse(aes(
    fill = run_id,
    color = run_id
    ),
    alpha = 0.05
    ) +
  scale_color_paletteer_d("ggsci::default_igv") +
  scale_fill_paletteer_d("ggsci::default_igv") +
  guides(color = guide_legend(nrow = 4)) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```

### UMAP by ethnicity
```{r UMAP by ethnicity, fig.width=12, fig.height=6}
p1 <-
  umap_results %>%
  select(umap_1, umap_2, sample_name, ethnicity) %>%
  mutate(ethnicity = as_factor(ethnicity)) %>%
  ggplot(
    aes(
      x = umap_1,
      y = umap_2,
      fill = ethnicity,
      line = 1
    ),
    shape = 1
  ) +
    geom_point(
      shape = 21,
      colour = "black",
      size = 2,
      stroke = 0.5,
      alpha = 0.6
    ) +
    scale_fill_paletteer_d(palette = "ggthemes::calc") +
    theme_cowplot() +
    theme(legend.position = "none")

p2 <-
  umap_results %>%
  select(umap_1, umap_2, sample_name, ethnicity) %>%
  mutate(ethnicity = as_factor(ethnicity)) %>%
  ggplot(
    aes(
      x = umap_1,
      y = umap_2,
      fill = ethnicity,
      color = ethnicity,
      line = 1
      ), shape = 1
    ) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
    ) +
  scale_fill_paletteer_d(palette = "ggthemes::calc") +
  scale_color_paletteer_d(palette = "ggthemes::calc") +
  theme_cowplot() +
  geom_mark_ellipse(
    alpha = 0.05
    ) +
  theme(legend.position = "bottom")

legend <- get_legend(p2)

plot_grid(
  plot_grid(
    p1 + theme(legend.position = "none"),
    p2 + theme(legend.position = "none")
    ),
  legend,
  nrow = 2,
  rel_heights = c(1, .2)
)
```

### UMAP, by initial sample concentration
```{r, fig.width=8, fig.height=6}
ggplot(
  data = umap_results,
  mapping = aes(
    x = umap_1,
    y = umap_2,
    fill = initial_concentration_ng_ul
  )
) +
  geom_point(
    shape = 21,
    colour = "black",
    size = 2,
    stroke = 0.5,
    alpha = 0.6
  ) +
  scale_fill_gradient2(low = "blue",
                       mid = "grey90",
                       high = "red",
                       midpoint = 10)
```

### UMAP by K-means cluster
```{r UMAP by K-means cluster, fig.height=6, fig.width=7}
umap_results %>%
  plot_ly(
    x = ~umap_1,
    y = ~umap_2,
    color = ~cluster,
    alpha = 0.75,
    
    colors = group_pal$cluster,
    text = str_glue("Sample: {umap_results$sample_name}<br>Classification: {umap_results$disease_class}<br>Ethnicity {umap_results$ethnicity}<br>Cluster: {umap_results$cluster}<br>Sex: {umap_results$sex}<br>Plate: {umap_results$run_id}")
  ) %>%
  add_markers(
    text = str_glue("Sample: {umap_results$sample_name}<br>Classification: {umap_results$disease_class}<br>Ethnicity {umap_results$ethnicity}<br>Cluster: {umap_results$cluster}<br>Sex: {umap_results$sex}<br>Plate: {umap_results$run_id}"),
    color = ~cluster,
    size = 2.5,
    alpha = 0.9,
    hovertemplate = "%{text}",
    marker = list(
      size = 10,
      line = list(
        color = 'rgba(0, 0, 0, .8)',
        width = 1.5
      )
    )
  )
```

### UMAP in 3D by K-means cluster
```{r UMAP in 3D by K-means cluster}
umap_results %>%
  plot_ly(
    x = ~umap_1,
    y = ~umap_2,
    z = ~umap_3,
    colors = group_pal$cluster
  ) %>%
  add_markers(
    text = str_glue("Sample: {umap_results$sample_name}<br>Classification: {umap_results$disease_class}<br>Ethnicity {umap_results$ethnicity}<br>Cluster: {umap_results$cluster}<br>Sex: {umap_results$sex}<br>Plate: {umap_results$run_id}"),
    color = ~cluster,
    size = 2.5,
    alpha = 0.9,
    hovertemplate = "%{text}",
    marker = list(
      size = 5,
      line = list(
        color = 'rgba(0, 0, 0, .8)',
        width = 1.5
      )
    )
  )
```

# Module scores {.tabset .tabset-fade .tabset-pills}

## SLE Interferon module scores
```{r SLE Interferon module scores, fig.height=9, fig.width=4}
loadd(module_scores_with_viral, ifn_modules)
range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name, one_of(ifn_modules)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## SLE Inflammatory module scores
```{r SLE Inflammatory module scores, fig.height=9, fig.width=6}
loadd(inflame_modules)
range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name, one_of(inflame_modules)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## Low-density granulocyte module scores
```{r Low-density granulocyte module scores, fig.height=9, fig.width=5}
loadd(ldg_modules)
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(names(ldg_modules))) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    angle = 315,
    border_color = NA)
```

## SLE Modules with an annotated associated pathway
```{r SLE Modules with an annotated associated pathway, fig.height=9, fig.width=11}
loadd(annotated_modules)

range_bound = 4
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(annotated_modules$module)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale = 'column',
    fontsize = 9,
    show_rownames = FALSE,
    annotation_row = annotation_info,
    annotation_col = column_to_rownames(annotated_modules, var = "module"),
    annotation_colors = group_pal,
    angle_col = 315,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA)

```

## All SLE modules
```{r All SLE modules, fig.height=9, fig.width=11}
loadd(module_annot)
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         matches("^M[[:digit:]]+")) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(scale = 'column',
           fontsize = 9,
           annotation_row = annotation_info,
           annotation_col = module_annot,
           annotation_colors = group_pal %>%
             `$<-`(type,
                   c(.$type,
                     Undetermined = "#000000")),
           show_rownames = FALSE,
           show_colnames = FALSE,
           angle_col = 315,
           breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
  border_color = NA)
```

## Expression of the annotated Banchereau modules {.tabset .tabset-fade .tabset-pills}

### By cluster
```{r Banchereau modules by cluster, echo=FALSE, fig.height=8, fig.width=12, warning=FALSE}
loadd(module_scores_with_viral)

annotated_module_scores <- 
  module_scores_with_viral %>%
  mutate(cluster = as_factor(cluster)) %>%
  select(cluster,
         disease_class,
         one_of(annotated_modules$module))

annotated_mod_list <-
  annotated_modules %>%
  mutate(
    module_type = paste(module,
                        type,
                        sep=" - ")
  ) %>%
  select(-type) %>%
  deframe()

annotated_module_scores <-
  names(annotated_module_scores) %>%
  recode(!!!annotated_mod_list) %>%
  setNames(nm = .,
           object = annotated_module_scores)

number_of_modules <-
  ncol(annotated_module_scores) - 1

number_of_pages <-
  ceiling(number_of_modules/9)


annotated_module_scores_pivot <-
  annotated_module_scores %>%
  pivot_longer(
    cols = starts_with("M"),
    names_to = "module",
    values_to = "score"
  )

annotated_module_stats_by_cluster <-
  annotated_module_scores_pivot %>%
  group_by(module) %>%
  wilcox_test(
    score ~ cluster,
    p.adjust.method = "BH"
  )

annotated_module_stats_by_cluster_with_spacing <-
  filter(annotated_module_stats_by_cluster,
         p.adj.signif != "ns") %>%
  add_xy_position(step.increase = 0.05)
  
annotated_module_cluster_graphs <-
  annotated_module_scores_pivot %>%
  select(-disease_class) %>%
  mutate(cluster = as_factor(cluster)) %>%
  ggviolin(x = "cluster",
           y = "score",
           fill = "cluster",
           draw_quantiles = c(0.25, 0.5, 0.75)) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  geom_jitter(size = 1, alpha = 0.5) +
  stat_pvalue_manual(
    data = annotated_module_stats_by_cluster_with_spacing,
    label = "p.adj.signif",
    hide.ns = TRUE,
    bracket.size = 0.1,
    tip.length = 0.01,
    size = 3
  ) +
  theme_cowplot() +
  theme(
    axis.text.x =
      element_text(
        angle = 45,
        hjust = 1,
        vjust = 1,
        size = 9
        ),
    axis.text.y =
      element_text(size = 9)
    )
  

for(i in seq(number_of_pages)){
  print(
    annotated_module_cluster_graphs +
      facet_wrap_paginate(
        facets = vars(module),
        scales = "free_y",
        nrow = 2,
        ncol = 3,
        page = i)
  )
}
```

### By disease classification
```{r Banchereau modules by disease, echo=FALSE, fig.height=8, fig.width=12, warning=FALSE}

annotated_module_stats_by_disease <-
  annotated_module_scores_pivot %>%
  group_by(module) %>%
  wilcox_test(
    score ~ disease_class,
    p.adjust.method = "BH"
  )

annotated_module_stats_by_disease_class_with_spacing <-
  filter(annotated_module_stats_by_disease,
         p.adj.signif != "ns") %>%
  add_xy_position(step.increase = 0.05)
  
annotated_module_disease_class_graphs <-
  annotated_module_scores_pivot %>%
  select(-cluster) %>%
  mutate(disease_class = as_factor(disease_class)) %>%
  ggviolin(x = "disease_class",
           y = "score",
           fill = "disease_class",
           draw_quantiles = c(0.25, 0.5, 0.75)
           ) +
  geom_jitter(size = 1, alpha = 0.5) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  stat_pvalue_manual(
    data = annotated_module_stats_by_disease_class_with_spacing,
    label = "p.adj.signif",
    hide.ns = TRUE,
    bracket.size = 0.1,
    tip.length = 0.01,
    size = 3
  ) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 9),
        axis.text.y = element_text(size = 9))
  
for(i in seq(number_of_pages)){
  print(
    annotated_module_disease_class_graphs +
      facet_wrap_paginate(
        facets = vars(module),
        scales = "free_y",
        nrow = 2,
        ncol = 3,
        page = i)
  )
}
```

# Viral transcription {.tabset .tabset-fade .tabset-pills}

Samples were examined for transcripts from the following viruses:
HSV-1, HSV-2, VZV, EBV, EBV-2, HCMV, HHV7, KSHV, Adenovirus 2, B19, HIV-1, HBV, HPV-2, HPV-16, JCV, and MCV-1

## Detected viral transcription and Interferon module scores
```{r fig.height=10, fig.width=10}
loadd(detected_viral_transcripts)
range_bound = 5
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(ifn_modules,
                detected_viral_transcripts)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(
    scale="column",
    fontsize = 9,
    show_rownames = FALSE,
    annotation_colors = group_pal,
    annotation_row = annotation_info,
    breaks = zscore_range,
    color = viridis(n = length(zscore_range)-1, option = "D"),
    border_color = NA,
    angle=315)
```

## Detected viral transcription and Inflammatory module scores
```{r fig.height=10, fig.width=9}
range_bound = 5
zscore_range <- c(-range_bound:range_bound)

module_scores_with_viral %>%
  select(sample_name,
         one_of(inflame_modules,
                detected_viral_transcripts)) %>%
  column_to_rownames("sample_name") %>%
  pheatmap(scale = "column",
           annotation_colors = group_pal,
           annotation_row = annotation_info,
           breaks = zscore_range,
           fontsize = 9,
           show_rownames = FALSE,
           color = viridis(n = length(zscore_range)-1, option = "D"),
           border_color = NA,
           angle=315)
```

# WGCNA

## Expression similarity dendrogram
```{r WGCNA Expression similarity dendrogram, echo=FALSE, fig.height=6, fig.width=10}
loadd(wgcna_modules)
plotDendroAndColors(dendro = wgcna_modules$dendrograms[[1]],
                    colors = wgcna_modules$colors[wgcna_modules$blockGenes[[1]]],
                    groupLabels = "Module colors",
                    main = "Gene dendrogram and module colors in block 1",
                    dendroLabels = FALSE,
                    hang = 0.03,
                    addGuide = TRUE,
                    guideHang = 0.05)
```

## Expression of identified modules {.tabset .tabset-fade .tabset-pills}

### By cluster
```{r wgcna modules by cluster, echo=FALSE, fig.height=8, fig.width=12}
number_of_modules <-
  module_scores_with_viral %>%
  select(matches("^ME"))%>%
  ncol()

number_of_pages <-
  ceiling(number_of_modules/9)

comparisons <- unlist(
  apply(
    X = combinations(n = length(levels(module_scores_with_viral$cluster)),
                     r = 2,
                     v =  levels(module_scores_with_viral$cluster)),
    MARGIN = 1,
    FUN =  list),
  recursive = FALSE)

module_cluster_graphs <-
  module_scores_with_viral %>%
  select(cluster,
         matches("^ME")) %>%
  pivot_longer(-cluster,
               names_to = "module",
               values_to = "score") %>%
  mutate(cluster = as_factor(cluster)) %>%
  ggplot(aes(x = cluster,
             y = score,
             fill = cluster)) +
  geom_violin() +
  geom_jitter(size = 0.5, alpha = 0.25) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  stat_compare_means(
    comparisons = comparisons,
    hide.ns = TRUE,
    method = "wilcox",
    label= "p.adj",
    size = 3,
    vjust=0.2
  )

for(i in seq(number_of_pages)){
  print(
    module_cluster_graphs +
      facet_wrap_paginate(
        facets = vars(module),
        scales = "free_y",
        nrow = 3,
        ncol = 3,
        page = i) +
      theme_cowplot()
  )
}
```

### By disease classification
```{r wgcna by disease, echo=FALSE, fig.height=8, fig.width=12}
comparisons <- unlist(
  apply(
    X = combinations(n = length(levels(module_scores_with_viral$disease_class)),
                     r = 2,
                     v = levels(module_scores_with_viral$disease_class)),
    MARGIN = 1,
    FUN =  list),
  recursive = FALSE)

module_cluster_graphs <-
  module_scores_with_viral %>%
  select(disease_class,
         matches("^ME")) %>%
  pivot_longer(-disease_class,
               names_to = "module",
               values_to = "score") %>%
  mutate(disease_class = as_factor(disease_class)) %>%
  ggplot(aes(x = disease_class,
             y = score,
             fill = disease_class)) +
  geom_violin() +
  geom_jitter(size = 0.5, alpha = 0.25) +
  scale_fill_paletteer_d("ggsci::uniform_startrek") +
  stat_compare_means(
    comparisons = comparisons,
    hide.ns = TRUE,
    method = "wilcox",
    label= "p.adj",
    size = 3,
    vjust=0.2
  )

for(i in seq(number_of_pages)){
  print(
    module_cluster_graphs +
      facet_wrap_paginate(
        facets = vars(module),
        scales = "free_y",
        nrow = 3,
        ncol = 3,
        page = i) +
      theme_cowplot()
  )
}
```

## WGCNA eigenvalues
```{r wgcna eigenvalues, fig.height=9, fig.width=11}
range_bound = 3
zscore_range <- c(-range_bound:range_bound)

tmp <-
  module_scores_with_viral %>%
  select(sample_name,
         matches("^ME")) %>%
  column_to_rownames("sample_name")
  
cluster_counts <-
  annotation_info %>%
  group_by(cluster) %>%
  tally() %>%
  pull(n) %>%
  purrr::accumulate(.f = sum)

cluster_sample_order <-
  annotation_info %>%
  as_tibble(rownames = "sample_name") %>%
  arrange(cluster) %>%
  pull(sample_name)


tmp[cluster_sample_order,] %>%
  pheatmap(scale = 'column',
           fontsize = 9,
           cluster_rows = FALSE,
           gaps_row = cluster_counts,
           annotation_row = annotation_info,
           annotation_colors = group_pal %>%
             `$<-`(type,
                   c(.$type,
                     Undetermined = "#000000")),
           show_rownames = FALSE,
           show_colnames = TRUE,
           angle_col = 315,
           breaks = zscore_range,
           color = viridis(n = length(zscore_range)-1, option = "D"),
           border_color = NA)
```

## Gene ontology GSEA of WGCNA modules
```{r wgcna gsea, fig.height=11, fig.width=12}
loadd(MEplotting)

for(i in 1:ceiling(length(unique(MEplotting$module))/6)){
    
      plots <-
        ggplot(data = MEplotting,
               mapping = aes(
                 y = GeneRatio,
                 x = order,
                 color = p.adjust)
               ) +
          geom_point(stat = "identity",
                     size = 2) +
          scale_x_continuous(breaks = MEplotting$order,
                           labels = str_wrap(MEplotting$ID, width = 40),
                           expand = c(0.1,0.1)) +
          scale_color_gradient(low = "blue", high = "red") +
          labs(color = "adjusted\np value") +
          facet_wrap_paginate(~ module,
                              ncol = 2,
                              nrow = 3,
                              scales = "free",
                              page = i) +
          theme(axis.text.y = element_text(size = 9),
                axis.text.x = element_text(size = 9),
                panel.background = element_rect(fill = "white", linetype = "solid"),
                panel.border = element_rect(colour = "black", fill=NA, size=1),
                panel.grid = element_line(color = "grey90"),
                axis.title.y = element_blank()) +
          coord_flip()
      print(plots)
}
```


## Use of WGCNA modules in classifying clusters.

The module eigengene scores were used to train a random forest model with repeated cross validation to predict cluster identity.

The model was trained with 75% of the dataset and tested on 25%.

Note: Subjects with LP were excluded due to the small population size.

```{r wgcna cluster confusion}
loadd(wgcna_cluster_rf_cv, wgcna_cluster_test)

caret::confusionMatrix(
  data = predict(wgcna_cluster_rf_cv, wgcna_cluster_test),
  reference = as_factor(wgcna_cluster_test$cluster)
)
```

The relative importance of each eigengene in classification:

```{r wgcna cluster importance}
loadd(wgcna_cluster_rf_cv_varImp)

wgcna_cluster_rf_cv_varImp
```

## Use of WGCNA modules in labeling disease class.

The module eigengene scores were used to train a random forest model with repeated cross validation to predict cluster identity.

The model was trained with 75% of the dataset and tested on 25%.

```{r wgcna disease confusion}
loadd(wgcna_disease_class_rf_cv, wgcna_disease_class_test)

caret::confusionMatrix(
  data = predict(wgcna_disease_class_rf_cv, wgcna_disease_class_test),
  reference = as_factor(wgcna_disease_class_test$disease_class)
)
```

The relative importance of each eigengene in classification:

```{r wgcna disease importance}
loadd(wgcna_disease_class_rf_cv_varImp)

wgcna_disease_class_rf_cv_varImp
```

## Use of module modules in classifying clusters.

The module eigengene scores were used to train a random forest model with repeated cross validation to predict cluster identity.

The model was trained with 75% of the dataset and tested on 25%.

```{r Banchereau cluster confusion}
loadd(module_cluster_rf_cv, module_cluster_test)

caret::confusionMatrix(
  data = predict(module_cluster_rf_cv, module_cluster_test),
  reference = as_factor(module_cluster_test$cluster)
)
```

The relative importance of each eigengene in classification:

```{r Banchereau cluster importance}
loadd(module_cluster_rf_cv_varImp)

module_cluster_rf_cv_varImp
```

## Use of module modules in labeling disease class.

The module eigengene scores were used to train a random forest model with repeated cross validation to predict cluster identity.

The model was trained with 75% of the dataset and tested on 25%.

```{r Banchereau disease confusion}
loadd(module_disease_class_rf_cv, module_disease_class_test)

caret::confusionMatrix(
  data = predict(module_disease_class_rf_cv, module_disease_class_test),
  reference = as_factor(module_disease_class_test$disease_class)
)
```

The relative importance of each eigengene in classification:

```{r Banchereau disease importance}
loadd(module_disease_class_rf_cv_varImp)

module_disease_class_rf_cv_varImp
```

```{r steroid-use correlation}
loadd(cd72_and_steroids, cd72_prednisone_correlation, cd72_depomedrol_correlation)
```

```{r cd72/prednisone}
cd72_and_steroids %>%
  ggplot(aes(x = prednisone, y = CD72, color = disease_class)) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = lm) +
  stat_cor() +
  theme_pubr()
```

```{r cd72/depomedrol}
cd72_and_steroids %>%
  ggplot(aes(x = depomedrol_methylprednisone, y = CD72, color = disease_class)) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = lm) +
  stat_cor() +
  theme_pubr()

```{r}
sessioninfo::session_info()
```
